<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Marcin Kierczak" />


<title>Best Coding Practices – tidy, neat and efficient code</title>

<script src="lab_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="lab_files/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="lab_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="lab_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="lab_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="lab_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="lab_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="lab_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="lab_files/navigation-1.1/tabsets.js"></script>
<script src="lab_files/navigation-1.1/codefolding.js"></script>
<link href="lab_files/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="lab_files/highlightjs-9.12.0/highlight.js"></script>
<link href="lab_files/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="lab_files/pagedtable-1.1/js/pagedtable.js"></script>
<link href="lab_files/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="lab_files/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="lab_files/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Best Coding Practices – tidy, neat and efficient code</h1>
<h3 class="subtitle">RaukR 2019 • Advanced R for Bioinformatics</h3>
<h4 class="author">Marcin Kierczak</h4>

</div>


<p><link href="https://fonts.googleapis.com/css?family=Roboto|Source+Sans+Pro:300,400,600|Ubuntu+Mono&amp;subset=latin-ext" rel="stylesheet"></p>
<p><img src="assets/logo.svg" alt="logo_raukr" class="trlogo"></p>
<p><img src="assets/logo.svg" alt="logo_raukr" class="trlogo"></p>
<hr />
<p class="abstract">
This is the Tidyverse course work material for RaukR 2019.
</p>
<hr />
<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>Welcome to the hands-on workshop “Tidy Work in Tidyverse”. Most of the things necessary to complete the tutorials and challenges were covered in the lecture. However, sometimes the tasks require that you check the docs or search online. Not all our solutions are optimal. Let us know if you can do better or solve things in a different way. If stuck, look at hints, next google and if still stuck, turn to TA. It is a lot of material, do not fee bad if you do not solve all tasks. If you completed Challenge 3 you are good and have used the most important features of tidyverse! Good luck!</p>
</div>
<div id="general-excercises" class="section level1">
<h1><span class="header-section-number">2</span> General excercises</h1>
<p>Datasets are available <a href="https://www.dropbox.com/s/dxxt959g6iefeaf/tidyverse_lab_data.zip?dl=0">here</a>.</p>
<div id="pipes" class="section level2">
<h2><span class="header-section-number">2.1</span> Pipes</h2>
<p>Rewrite the following code chunks as pipes (Load package <code>magrittr</code> because <code>tidyverse</code> supports only the <code>%&gt;%</code> pipe!):</p>
<div id="chunk-1" class="section level3">
<h3><span class="header-section-number">2.1.1</span> Chunk 1</h3>
<pre><code>my_cars &lt;- mtcars[, c(1:4, 10)]
my_cars &lt;- my_cars[my_cars$disp &gt; mean(my_cars$disp), ]
my_cars &lt;- colMeans(my_cars)</code></pre>
<pre class="r"><code>mtcars %&gt;%
  select(c(1:4, 10)) %&gt;%
  filter(disp &gt; mean(disp)) %&gt;%
  colMeans() -&gt; my_cars</code></pre>
<p>What is wrong with our solution?</p>
<pre class="r"><code># It is better to have the result assigned on the left hand side: result &lt;- expression. In this case the expression is the **whole** pipe.
# Our &#39;expression -&gt; result&#39; is correct but can easily be missed when reading the code.</code></pre>
</div>
<div id="chunk-2" class="section level3">
<h3><span class="header-section-number">2.1.2</span> Chunk 2</h3>
<p>The <code>summary(x)</code> function is a bit special: when you type <code>summary(x)</code> in the console, <code>print</code> is called in an implicite way. Pipe call does not do, so you will have to invoke <code>print</code> in an explicite way. But the <code>%T&gt;%</code> does unbranch for one call only, you will have to make printing of the <code>summary</code> a one single composed call using <code>{}</code>.</p>
<pre><code>summary(cars)
colSums(cars)</code></pre>
<pre class="r"><code>cars %T&gt;% {print(summary(.))} %&gt;% colSums()</code></pre>
</div>
<div id="chunk-3" class="section level3">
<h3><span class="header-section-number">2.1.3</span> Chunk 3</h3>
<p>Rewrite correlations to pipes.</p>
<pre><code>cor(mtcars$gear, mtcars$mpg)</code></pre>
<pre class="r"><code>mtcars %$% cor(gear, mpg)</code></pre>
<pre><code>cor(mtcars)</code></pre>
<pre class="r"><code>mtcars %&gt;% cor()</code></pre>
</div>
<div id="chunk-4" class="section level3">
<h3><span class="header-section-number">2.1.4</span> Chunk 4</h3>
<p>Given is the <code>dim_summary(nrows, ncols)</code> function which takes matrix <code>nrows</code> and <code>ncols</code> as arguments and prints this info about its dimensions:</p>
<pre><code>dim_summary &lt;- function(nrows, ncols) {
  print(
    paste0(&#39;Matrix M has: &#39;, nrows, &#39; rows and &#39;, ncols, &#39; columns.&#39;)
  )
}</code></pre>
<p>Rewrite the code chunks below as pipes:</p>
<pre><code>distr1 &lt;- rnorm(16)
M &lt;- matrix(distr1, ncol = 4)
plot(M)
M &lt;- M + sample(M)
dim_summary(nrows = nrow(M), ncols = ncol(M))</code></pre>
<pre><code>distr2 &lt;- rnorm(16)
N &lt;- matrix(distr2, ncol = 4)
colnames(N) &lt;- (letters[1:4])
summary(N)</code></pre>
<pre><code>P &lt;- M %x% t(N)
heatmap(P)
colnames(P) &lt;- letters[1:dim(P)[2]]
cor(P[ ,&#39;a&#39;], P[ ,&#39;i&#39;])</code></pre>
<p>A class of functions, called the <em>replacement functions</em> are of the form <code>function(arguments)&lt;-value</code> and <code>rownames(x) &lt;- c('a', 'b', 'c')</code> is a good example of a replacement function. When writing pipes, we have bear in mind that whole <code>function&lt;-</code> is the name of the replacement function and thus we have to use it as such in the pipe.</p>
<p><strong>Note.</strong> Not always one can put everything into one single pipe, sometimes the results of running two or more pipes have to be used in the final pipe.</p>
<pre class="r"><code>dim_summary &lt;- function(nrows, ncols) {
  print(paste0(&#39;Matrix M has: &#39;, nrows, &#39; rows and &#39;, ncols, &#39; columns.&#39;))
}

M &lt;- rnorm(16) %&gt;%
  matrix(ncol = 4) %T&gt;%
  plot() %&gt;%
  `+`(., sample(.)) %T&gt;%
  {dim_summary(nrow(.), ncol(.))}

N &lt;- rnorm(16) %&gt;%
  matrix(ncol = 4) %&gt;%
  `colnames&lt;-`(letters[1:4]) %&gt;%
  summary() 

P &lt;- M %&gt;%
  `%x%`(., N) %T&gt;%
  heatmap() %&gt;%
  `colnames&lt;-`(letters[1:dim(.)[2]]) %&gt;%
  as_data_frame() %$%
  cor(a, i)</code></pre>
</div>
</div>
<div id="tibbles" class="section level2">
<h2><span class="header-section-number">2.2</span> Tibbles</h2>
<div id="task-1" class="section level3">
<h3><span class="header-section-number">2.2.1</span> Task 1</h3>
<ul>
<li>Convert the <code>mtcars</code> dataset to a tibble <code>vehicles</code>.</li>
<li>Select the number of cylinders (<code>cyl</code>) variable using:
<ul>
<li>the <code>[[index]]</code> accessor,</li>
<li>the <code>[[string]]</code> accessor,</li>
<li>the <code>$</code> accessor.</li>
</ul></li>
<li>Do the same selection as above, but using pipe and placeholders (use all three ways of accessing a variable).</li>
<li>Print the tibble.</li>
<li>Print the 30 first rows of the tibble.</li>
<li>Change the default behaviour of printing a tibble so that at least 15 and at most 30 rows are printed.</li>
<li>What is the difference between the <code>tibble.print_max</code> and <code>dplyr.print_min</code>? Is there any? Test it.</li>
<li>Convert <code>vehicles</code> back to a <code>data.frame</code> called <code>automobiles</code>.</li>
</ul>
<pre class="r"><code># 1
vehicles &lt;- mtcars %&gt;% as_tibble()

# 2
vehicles[[&#39;cyl&#39;]]
vehicles[[2]]
vehicles$cyl

# 3
vehicles %T&gt;%
  {print(.[[&#39;cyl&#39;]])} %T&gt;%
  {print(.[[2]])} %&gt;%
  .$cyl

# 4
vehicles

# 5
vehicles %&gt;% head(n = 30)

# 6
options(tibble.print_min = 15, tibble.print_max = 30)

# 7
# In theory there should be no difference. dplyr imports tibble from the tibble package
# and dplyr.width, dplyr.print_min and dplyr.print_min are passed down to the tibble.
# But test both behaviours. First with only the tibble package loaded, later with dplyr # loaded.

# 8
automobiles &lt;- as.data.frame(vehicles)</code></pre>
</div>
<div id="task-2" class="section level3">
<h3><span class="header-section-number">2.2.2</span> Task 2</h3>
<p>Create the following tibble using <code>tribble()</code>:</p>
<table>
<thead>
<tr class="header">
<th align="right">id</th>
<th align="left">event</th>
<th align="left">date</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="left">success</td>
<td align="left">24-04-2017</td>
</tr>
<tr class="even">
<td align="right">2</td>
<td align="left">failed</td>
<td align="left">25-04-2017</td>
</tr>
<tr class="odd">
<td align="right">3</td>
<td align="left">failed</td>
<td align="left">25-04-2017</td>
</tr>
<tr class="even">
<td align="right">4</td>
<td align="left">success</td>
<td align="left">27-04-2017</td>
</tr>
</tbody>
</table>
<pre class="r"><code>tab &lt;- tribble(
  ~id, ~event, ~date,
  1, &#39;success&#39;, &#39;24-04-2017&#39;,
  2, &#39;failed&#39;, &#39;25-04-2017&#39;,
  3, &#39;failed&#39;, &#39;25-04-2017&#39;,
  4, &#39;success&#39;, &#39;27-04-2017&#39;
)</code></pre>
</div>
<div id="task-3" class="section level3">
<h3><span class="header-section-number">2.2.3</span> Task 3</h3>
<p>Compare the performance of <code>as.data.frame()</code>, <code>as_data_frame()</code> and <code>as_tibble()</code> on a
100 x 30 matrix filled with random integers. Use package <code>microbenchmark</code>. Fill in your result <a href="https://docs.google.com/spreadsheets/d/1_2tDeEkDVS06RkB437yBI1XEB5SUebtHWyxAf_aRJu4/edit#gid=99106509">here</a> in the Tidyverse Lab sheet, Tibbles – performance.</p>
<pre class="r"><code>tst &lt;- replicate(30, sample(100), simplify = TRUE)
colnames(tst) = paste0(rep(&#39;col&#39;, times = dim(tst)[2]), 1:dim(tst)[2])
microbenchmark::microbenchmark(
  as.data.frame(tst),
  as_data_frame(tst),
  as_tibble(tst)
)</code></pre>
</div>
<div id="task-4" class="section level3">
<h3><span class="header-section-number">2.2.4</span> Task 4</h3>
<p>Do you think tibbles are lazy? Try to create a tibble that tests whether <em>lazy evaluation</em> applies to tibbles too.</p>
<pre class="r"><code>tibble(x = sample(1:10, size = 10, replace = T), y = log10(x))</code></pre>
</div>
</div>
<div id="parsing" class="section level2">
<h2><span class="header-section-number">2.3</span> Parsing</h2>
<p>Parse the following vectors:</p>
<ul>
<li><code>vec1 &lt;- c(1, 7.2, 3.84, -5.23)</code> – parse it as double (any problems? why?).</li>
<li>now parse the same vector <code>c(1, 7.2, 3.84, -5.23)</code> as integer. What happens?</li>
<li>Can you still parse it as integer somehow?</li>
<li>Parse as double <code>vec2 &lt;- c('2', '3,45', '?', '-7,28')</code></li>
<li>Parse correctly <code>vec3 &lt;- c('2', '3,45', '?', '-7.28')</code></li>
<li>Parse the following guessing the parser: <code>vec4 &lt;- c('barrel: 432.7$', 'liter: 15.42PLN', 'gallon costs approx 32.1SEK', 'sunny, wind gusts up till 55m/s')</code></li>
<li>Can you parse <code>vec4</code> as number? Do it if you can.</li>
<li>Parse <code>vec5 &lt;- &quot;25 Dec 2015&quot;</code> as date (hint: <code>?parse_date()</code>).</li>
<li>Parse <code>10_Jul_1410</code> as date.</li>
</ul>
<pre class="r"><code>vec1 &lt;- c(1, 7.2, 3.84, -5.23)
vec2 &lt;- c(&#39;2&#39;, &#39;3,45&#39;, &#39;?&#39;, &#39;-7,28&#39;)
vec3 &lt;- c(&#39;2&#39;, &#39;3,45&#39;, &#39;?&#39;, &#39;-7.28&#39;)
vec4 &lt;- c(&#39;barrel: 432.7$&#39;, &#39;liter: 15.42PLN&#39;, &#39;gallon costs approx 32.1SEK&#39;, &#39;sunny, wind gusts up till 55m/s&#39;)
vec5 &lt;- &quot;25 Dec 2015&quot;
parse_double(vec1)
parse_integer(vec1)
parse_integer(as.integer(vec1)) # Is it the best way? Hint: rounding.
parse_double(vec2, na = &#39;?&#39;, locale = locale(decimal_mark = &#39;,&#39;))
parse_number(vec2, na = &#39;?&#39;, locale = locale(decimal_mark = &#39;,&#39;))
guess_parser(vec4)
parse_guess(vec4)
# Yes, you can:
parse_number(vec4)
parse_date(vec5, format=&quot;%d %b %Y&quot;)
parse_date(&quot;10_Jul_1410&quot;, format=&quot;%d%.%b%.%Y&quot;)</code></pre>
</div>
</div>
<div id="nyc-flights-challenge" class="section level1">
<h1><span class="header-section-number">3</span> NYC flights Challenge</h1>
<p>The <code>nycflights13</code> package contains information about all flights that departed from NYC (i.e., EWR, JFK and LGA) in 2013: 336,776 flights with 16 variables. To help understand what causes delays, it also includes a number of other useful datasets: weather, planes, airports, airlines. We will use it to train working with tibbles and <code>dplyr</code>.</p>
<div id="task-1-selecting-column" class="section level2">
<h2><span class="header-section-number">3.1</span> Task 1 – selecting column</h2>
<ul>
<li>load the <code>nycflights13</code> package (install if necessary),</li>
<li>read about the data in the package docs,</li>
<li>inspect the <code>flights</code> tibble.</li>
<li>select all columns but <code>carrier</code> and <code>arr_time</code>,</li>
<li>select <code>carrier</code>, <code>tailnum</code> and <code>origin</code>,</li>
<li>hide columns from <code>day</code> through <code>carrier</code>,</li>
<li>select all columns that have to do with <code>arr</code>ival (hint: <code>?tidyselect</code>),</li>
<li>select columns based on a vector <code>v &lt;- c(&quot;arr_time&quot;, &quot;sched_arr_time&quot;, &quot;arr_delay&quot;)</code>,</li>
<li>rename column <code>dest</code> to <code>destination</code> using:
<ul>
<li><code>select()</code> and</li>
<li><code>rename()</code><br />
What is the difference between the two approaches?</li>
</ul></li>
</ul>
<pre class="r"><code>install.packages(&#39;nycflights13&#39;)

library(&#39;nycflights13&#39;)

?nycflights13

flights

flights %&gt;% select(-carrier, -arr_time)

flights %&gt;% select(carrier, tailnum, origin)

flights %&gt;% select(-(day:carrier))

flights %&gt;% select(contains(&#39;arr_&#39;)) # or

v &lt;- c(&quot;arr_time&quot;, &quot;sched_arr_time&quot;, &quot;arr_delay&quot;)
flights %&gt;% select(v) # or
flights %&gt;% select(one_of(v))

flights %&gt;% select(destination = dest)
flights %&gt;% rename(destination = dest)
# select keeps only the renamed column while rename returns the whole dataset
# with the column renamed</code></pre>
</div>
<div id="task-2-filtering-rows" class="section level2">
<h2><span class="header-section-number">3.2</span> Task 2 – filtering rows</h2>
<ul>
<li>filter only the flights that arrived ahead of schedule,</li>
<li>filter the flights that had departure delay between 10 and 33,</li>
<li>fish out all flights with unknown arrival time,</li>
<li>retrieve rows 1234:1258 (hint: <code>?slice</code>),</li>
<li>sample (<code>?sample_n()</code>) 3 random flights per day in March,</li>
<li>show 5 most departure-delayed flights in January per carrier,</li>
<li>retrieve all <code>unique()</code> routes and sort them by origin,</li>
<li>retrieve all <code>distinct()</code> routes and sort them by origin,</li>
<li>is <code>unique()</code> more efficient than <code>distinct()</code>?</li>
</ul>
<pre class="r"><code>flights %&gt;% filter(arr_delay &lt; 0)

flights %&gt;% filter(dep_delay &gt;= 10, dep_delay &lt;= 33) # or
flights %&gt;% filter(between(dep_delay, 10, 33))

flights %&gt;% filter(is.na(arr_time))

flights %&gt;% slice(1234:1258)

flights %&gt;% filter(month == 3) %&gt;%
  group_by(day) %&gt;%
  sample_n(3)

flights %&gt;%
  filter(month == 1) %&gt;%
  group_by(carrier) %&gt;%
  top_n(5, dep_delay)</code></pre>
</div>
<div id="task-3-transmutations" class="section level2">
<h2><span class="header-section-number">3.3</span> Task 3 – Trans(mutations)</h2>
<ul>
<li><p><code>air_time</code> is the amount of time in minutes spent in the air. Add a new column <code>air_spd</code> that will contain aircraft’s airspeed in mph,</p></li>
<li><p>as above, but keep only the new <code>air_spd</code> variable,</p></li>
<li><p>use <code>rownames_to_column()</code> on <code>mtcars</code> to add car model as an extra column,</p></li>
</ul>
<pre class="r"><code>flights %&gt;% mutate(air_spd = distance/(air_time / 60))

flights %&gt;% transmute(air_spd = distance/(air_time / 60))

mtcars %&gt;% rownames_to_column(&#39;model&#39;)</code></pre>
</div>
<div id="task-4-groups-and-counts" class="section level2">
<h2><span class="header-section-number">3.4</span> Task 4 – groups and counts</h2>
<ul>
<li>use <code>group_by()</code>, <code>summarise()</code> and <code>n()</code> to see how many planes were delayed (departure) every month,</li>
</ul>
<pre class="r"><code>flights %&gt;%
  filter(dep_delay &gt; 0) %&gt;%
  group_by(month) %&gt;%
  summarise(num_dep_delayed = n())</code></pre>
<ul>
<li>do the same but using <code>tally()</code> and <code>count()</code>,</li>
</ul>
<pre class="r"><code>flights %&gt;%
  filter(dep_delay &gt; 0) %&gt;%
  group_by(month) %&gt;%
  tally()

flights %&gt;%
  filter(dep_delay &gt; 0) %&gt;%
  count(month)</code></pre>
<ul>
<li>what was the mean <code>dep_delay</code> per month?</li>
</ul>
<pre class="r"><code>flights %&gt;%
  group_by(month) %&gt;%
  summarise(mean_dep_delay = mean(dep_delay, na.rm = T))</code></pre>
<ul>
<li>count the number of incoming delayed flights from each unique origin and sort origins by this count (descending),</li>
</ul>
<pre class="r"><code>flights %&gt;%
  filter(arr_delay &gt; 0) %&gt;%
  group_by(origin) %&gt;%
  summarise(cnt = n()) %&gt;%
  arrange(desc(cnt))</code></pre>
<ul>
<li>do the same using <code>tally()</code></li>
</ul>
<pre class="r"><code>flights %&gt;%
  filter(arr_delay &gt; 0) %&gt;%
  group_by(origin) %&gt;%
  tally(sort = T)</code></pre>
<ul>
<li>use <code>summarise()</code> to sum total <code>dep_delay</code> per month in hours,</li>
</ul>
<pre class="r"><code> flights %&gt;%
   group_by(month) %&gt;%
   summarize(tot_dep_delay = sum(dep_delay/60, na.rm = T))</code></pre>
<ul>
<li>use the <code>wt</code> parameter of <code>count()</code> (works with <code>tally()</code> too) to achieve the same,</li>
</ul>
<pre class="r"><code> flights %&gt;%
   group_by(month) %&gt;%
   count(wt = dep_delay/60)</code></pre>
<ul>
<li>run <code>group_size()</code> on <code>carrier</code> what does it return?</li>
</ul>
<pre class="r"><code>flights %&gt;%
    group_by(carrier) %&gt;%
    group_size()</code></pre>
<ul>
<li>use <code>n_groups()</code> to check the number of unique origin-carrier pairs,</li>
</ul>
<pre class="r"><code>flights %&gt;%
    group_by(carrier) %&gt;%
    n_groups()</code></pre>
<p><strong>Note on <code>ungroup</code></strong> Depending on the version of <code>dplyr</code> you may or may need to use the <code>ungroup()</code> if you want to group your data on some other variables:</p>
<pre><code>flights %&gt;%
  group_by(origin) %&gt;%
  mutate(mean_delay_orig = (mean(dep_delay, na.rm = T) + mean(arr_delay, na.rm = T)) / 2) %&gt;%
  ungroup() %&gt;%
  group_by(carrier) %&gt;%
  mutate(mean_delay_carr = (mean(dep_delay, na.rm = T) + mean(arr_delay, na.rm = T)) / 2) %&gt;%
  select(origin, carrier, mean_delay_orig, mean_delay_carr)</code></pre>
<p>may or may need ungroup depending on your <code>dplyr</code> version. In the newer versions, <code>summarise</code> and <code>mutate</code> drop one aggregation level.</p>
</div>
<div id="task-5-joins" class="section level2">
<h2><span class="header-section-number">3.5</span> Task 5 – joins</h2>
<p>Given the following tibbles:</p>
<p><code>set1</code>:</p>
<pre class="r"><code>knitr::kable(set1)</code></pre>
<table>
<thead>
<tr class="header">
<th align="left">id</th>
<th align="left">color</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">id1</td>
<td align="left">grey</td>
</tr>
<tr class="even">
<td align="left">id1</td>
<td align="left">red</td>
</tr>
<tr class="odd">
<td align="left">id2</td>
<td align="left">green</td>
</tr>
<tr class="even">
<td align="left">id3</td>
<td align="left">blue</td>
</tr>
</tbody>
</table>
<p>and <code>set2</code>:</p>
<pre class="r"><code>knitr::kable(set2)</code></pre>
<table>
<thead>
<tr class="header">
<th align="left">id</th>
<th align="left">size</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">id2</td>
<td align="left">XL</td>
</tr>
<tr class="even">
<td align="left">id3</td>
<td align="left">M</td>
</tr>
<tr class="odd">
<td align="left">id4</td>
<td align="left">M</td>
</tr>
</tbody>
</table>
<p>Perform joins on <code>id</code> that result in the grey area from the Venn diagrams below:</p>
<p><img src="lab_files/figure-html/unnamed-chunk-31-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<pre class="r"><code>left_join(set1, set2)</code></pre>
<p><img src="lab_files/figure-html/unnamed-chunk-33-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<pre class="r"><code>right_join(set1, set2)</code></pre>
<p><img src="lab_files/figure-html/unnamed-chunk-35-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<pre class="r"><code>inner_join(set1, set2) # or
semi_join(set1, set2) # semi_join removes duplicates in x
# and also returns only columns from x.</code></pre>
<p><img src="lab_files/figure-html/unnamed-chunk-37-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<pre class="r"><code>full_join(set1, set2) # or</code></pre>
<p><img src="lab_files/figure-html/unnamed-chunk-39-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<pre class="r"><code>anti_join(set1, set2) # or</code></pre>
</div>
</div>
<div id="tidying-data" class="section level1">
<h1><span class="header-section-number">4</span> Tidying data</h1>
<p>Now time to do some data tidying. First install a package with some untidy data:</p>
<pre><code>devtools::install_github(&quot;rstudio/EDAWR&quot;)
library(EDAWR)</code></pre>
<ul>
<li>tidy <code>cases</code> so that years are not in separate columns, but in the column called <code>year</code> containing a value per each year.</li>
</ul>
<pre class="r"><code>tidy_cases &lt;- cases %&gt;% gather(key = &#39;year&#39;, value = &#39;count&#39;, 2:4)</code></pre>
<ul>
<li>now time for the <code>pollution</code> dataset. Tidy it so that there separate columns for <code>large</code> and <code>small</code> pollution values.</li>
</ul>
<pre class="r"><code>tidy_pollution &lt;- pollution %&gt;% spread(size, amount)</code></pre>
<ul>
<li>the <code>storms</code> dataset contains the <code>date</code> column. Make it into 3 columns: <code>year</code>, <code>month</code> and <code>day</code>. Store the result as <code>tidy_storms</code></li>
</ul>
<pre class="r"><code>tidy_storms &lt;- storms %&gt;%
                          separate(col = date,
                                   into = c(&#39;year&#39;, &#39;month&#39;, &#39;day&#39;),
                                   sep = &#39;-&#39;)</code></pre>
<ul>
<li>now, merge <code>year</code>, <code>month</code> and <code>day</code> in <code>tidy_storms</code> into a <code>date</code> column again but in the “DD/MM/YYYY” format.</li>
</ul>
<pre class="r"><code>tidy_storms %&gt;% unite(col = &quot;date&quot;, 4:6, sep = &quot;/&quot;)</code></pre>
</div>
<div id="nanopore-channel-activity-challenge" class="section level1">
<h1><span class="header-section-number">5</span> Nanopore Channel Activity Challenge</h1>
<div id="introduction-1" class="section level2">
<h2><span class="header-section-number">5.1</span> Introduction</h2>
<p>You will be given a <code>fastq</code> file coming from MinION sequencer (Oxford Nanopore). This file contains test reads from the chicken genome. The flow-cell used here has 512 channels, each channel consists of 4 pores and only one pore is active at a time. Once your sequence gets stuck for some reason, the device will attempt to remove it from the pore by playing with reversing polarity on that pore. If this was successful, the pore will be re-used. Your task will be to visualise reading events from the meta-data in the <code>fastq</code> dataset and to see how each and every channel behaved. Also, you will plot the distribution of reading times.</p>
<p>Datasets are available <a href="https://www.dropbox.com/s/dxxt959g6iefeaf/tidyverse_lab_data.zip?dl=0">here</a>.</p>
</div>
<div id="preparations" class="section level2">
<h2><span class="header-section-number">5.2</span> Preparations</h2>
<p>First, we will need to load the necessary libraries. I will give you a hint – you need the following libraries:</p>
<ul>
<li><code>here</code> – not necessary, but it is an elegant way of reading the data locally from the project folder,</li>
<li><code>tidyverse</code> – well, quite obvious why,</li>
<li><code>ShortRead</code> from Bioconductor – to deal with short reads in <code>fastq</code>,</li>
<li><code>lubridate</code> – to figure out reading times.</li>
</ul>
<pre class="r"><code>library(here)
library(tidyverse)
library(ShortRead)
library(lubridate)</code></pre>
</div>
<div id="reading-data" class="section level2">
<h2><span class="header-section-number">5.3</span> Reading data</h2>
<p>Now, let’s read the fastq data. Check <code>ShortRead</code> documentation to see how to read our <code>fastq</code> file. Also, try to use package <code>here</code>. If you write: <code>data &lt;- here::here('data/my.fastq')</code>, the <code>my.fastq</code> file will be read from the <code>data</code> folder which is a subfolder of your project root, i.e. the folder where your r script is. It is a good practise and also prevents <a href="https://www.tidyverse.org/articles/2017/12/workflow-vs-script/">Jenny Bryan from coming to your office and setting your computer on fire</a>.</p>
<p>Now think a bit, to plot reading events, do we need al the data in the file or only some specific part? You may want to see some few first lines of the <code>fastq</code> to learn about the data structure.</p>
<pre class="r"><code>raw_data &lt;- here::here(&#39;docs/tidyverse_Marcin/lab/lab_assets/FUL1_fastqs_GRID2.fastq&#39;)
f &lt;- FastqFile(raw_data)
rfq &lt;- readFastq(f)
headers &lt;- rfq@id
close(f)</code></pre>
</div>
<div id="extracting-information-you-need" class="section level2">
<h2><span class="header-section-number">5.4</span> Extracting information you need</h2>
<p>In this step, we are extracting data from fastq headers of each and every read in the fastq file. Not super efficient and perhaps the slowest step of the whole analyses. Can you do it better than our example solution?</p>
<p>Desired output: a table (tibble/data.frame) with reads as rows and meta-data as columns.</p>
<p>Hint: use <code>strsplit()</code> to explode string data into columns and <code>str_remove_all()</code> to get rid of unwanted characters. Since we did not talk much about regular expressions and the <code>stringr</code> package:</p>
<ul>
<li>To split string on a particular character, group of characters use <code>strsplit</code>. Here we split on comma.</li>
</ul>
<pre><code>text &lt;- &quot;This text is long, or not?&quot;
strsplit(text, &#39;,&#39;)</code></pre>
<ul>
<li>To remove everything following a given character, e.g. comma:</li>
</ul>
<pre><code>str_remove_all(text, &quot;,.*&quot;)</code></pre>
<pre class="r"><code>data &lt;- dplyr::as_tibble(matrix(NA_character_, ncol = 6, nrow = length(headers)))
colnames(data) &lt;- c(&#39;id&#39;, &#39;run_id&#39;, &#39;sample_id&#39;, &#39;read&#39;, &#39;channel&#39;, &#39;start_time&#39;)
for (i in 1:length(headers)) {
  tmp &lt;- toString(headers[[i]])
  tmp_flat &lt;- unlist(strsplit(tmp, &#39; &#39;)) %&gt;%
    str_remove_all(&quot;.*=&quot;)
  data[i,] &lt;- tmp_flat
}</code></pre>
</div>
<div id="preparing-tidy-dataset" class="section level2">
<h2><span class="header-section-number">5.5</span> Preparing tidy dataset</h2>
<p>Now, the fun part begins:</p>
<ul>
<li>add column <code>start_dttm</code> that represents start time for a given read as proper <code>datetime</code> object (read <code>lubridate</code> docs) and</li>
<li>add column <code>chan</code> that is the proper numeric representation of the channel,</li>
<li>group reads by channel,</li>
<li>arrange them by time,</li>
<li>add time to next read (NA if this was the last read) and</li>
<li>sort by channel again.</li>
</ul>
<p>Hint: read about <code>lead()</code></p>
<pre class="r"><code>data2 &lt;- data %&gt;%
  mutate(start_dttm = as_datetime(start_time)) %&gt;%
  mutate(chan=as.numeric(channel)) %&gt;%
  group_by(chan) %&gt;%
  arrange(start_dttm) %&gt;%
  mutate(time_diff = lead(start_dttm) - start_dttm) %&gt;%
  arrange(chan)</code></pre>
</div>
<div id="plotting-events-per-channel" class="section level2">
<h2><span class="header-section-number">5.6</span> Plotting events per channel</h2>
<p><strong>Important!</strong> If you are not familiar with <code>ggplot</code> either:</p>
<ul>
<li>wait until tomorrow,</li>
<li>use base-R plotting, or</li>
<li>see our <code>ggplot2</code> solution.</li>
</ul>
<p>Here, we want to see what was happening in each channel over time. Plot the data you have just prepared so that:</p>
<ul>
<li>each point is the start of a new read,</li>
<li>colour corresponds to the lag to the next read.</li>
</ul>
<p>Can you visualise this in a better way? Different geometry?</p>
<pre class="r"><code>ggplot(data2, mapping = aes(x = start_dttm,
                            y = as.factor(chan),
                            col = as.numeric(time_diff)
                            )
       ) +
  geom_point(size = .5) +
  theme_bw()</code></pre>
<p><img src="lab_files/figure-html/plot_events-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</div>
<div id="distribution-of-time-intervals" class="section level2">
<h2><span class="header-section-number">5.7</span> Distribution of time intervals</h2>
<p>Now, we want to see how time-to-next-read is distributed. Since it has a veeeeery long right tail, I am cutting off everything above 200 seconds (just by eyeballing).</p>
<ul>
<li>plot time-to-next-read is distribution (you can use base-R <code>histogram</code>),</li>
<li>can you find a better cutoff value?</li>
</ul>
<pre class="r"><code># Show time-to-next read distribution
# thr &lt;- mean(data2$time_diff, na.rm = T) + 3 * sd(data2$time_diff, na.rm = T)
tmp &lt;- data2 %&gt;%
  ungroup() %&gt;%
  filter(time_diff &lt; 200) %&gt;%
  select(time_diff)

hist(as.numeric(tmp$time_diff), breaks = 1000, las=1)</code></pre>
<p><img src="lab_files/figure-html/time_to_next_1-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</div>
</div>
<div id="species-identification-challenge" class="section level1">
<h1><span class="header-section-number">6</span> Species Identification Challenge</h1>
<p>In this challenge, your task will be to analyse species composition of some samples. The samples, were actual products containing parts of plants. DNA has been isolated form the samples and an amplicon metabarcoding was performed using two sets of primers: for the ITS1 and the ITS2 region. Each sample had 3 technical replicates. Your task will be to transform BLAST output to a tidy form suitable for further analyses or visualisation.</p>
<div id="load-necessary-libraries" class="section level2">
<h2><span class="header-section-number">6.1</span> Load necessary libraries</h2>
<p>We will obviously need <code>tidyverse</code>, we will also do some string manipulations with <code>stringr</code> also <code>here</code> package is good to have.</p>
<pre class="r"><code>library(tidyverse)
library(stringr)
library(here)</code></pre>
</div>
<div id="input-variables" class="section level2">
<h2><span class="header-section-number">6.2</span> Input variables</h2>
<p>Here, we will define our input variables. We need:</p>
<ul>
<li><code>file</code> that contains the path to the dataset,</li>
<li><code>sample_name</code> is a string, the name of the sample you want to analyse,</li>
<li><code>threshold</code> is an intiger saying what is a the minimal number of replicates that have to contain an OTU in order to call it a true positive (TP),</li>
<li><code>strict</code> a logical. If set to TRUE, only the OTUs deemed TP will be shown.</li>
</ul>
<p>Below we set some examle values:</p>
<pre><code># Change the path to your project path, where your data is
file &lt;- here::here(&quot;docs/tidyverse_Marcin/lab/lab_assets/blast_result.csv&quot;)
sample_name &lt;- &#39;SAMPLE12&#39;
threshold &lt;- 1
strict &lt;- F</code></pre>
</div>
<div id="reading-the-data" class="section level2">
<h2><span class="header-section-number">6.3</span> Reading the data</h2>
<p>Now, you should read the data:</p>
<pre class="r"><code>species_orig &lt;- read_csv(file, col_names = c(&quot;sample&quot;,&quot;its&quot;,&quot;replicate&quot;,&quot;OTU&quot;,&quot;size&quot;,&quot;hit&quot;,&quot;perc_ident&quot;,&quot;score&quot;,&quot;family&quot;,&quot;species&quot;)) %&gt;% select(-score)</code></pre>
<table>
<thead>
<tr class="header">
<th align="left">sample</th>
<th align="left">its</th>
<th align="left">replicate</th>
<th align="left">OTU</th>
<th align="right">size</th>
<th align="left">hit</th>
<th align="right">perc_ident</th>
<th align="left">family</th>
<th align="left">species</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">SAMPLE10</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU1</td>
<td align="right">372</td>
<td align="left">KX522674</td>
<td align="right">97.08</td>
<td align="left">Anacardiaceae</td>
<td align="left">Spondias_tuberosa</td>
</tr>
<tr class="even">
<td align="left">SAMPLE10</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU1</td>
<td align="right">372</td>
<td align="left">AJ783644</td>
<td align="right">97.53</td>
<td align="left">Betulaceae</td>
<td align="left">Betula_populifolia</td>
</tr>
<tr class="odd">
<td align="left">SAMPLE10</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU1</td>
<td align="right">372</td>
<td align="left">AJ783641</td>
<td align="right">97.53</td>
<td align="left">Betulaceae</td>
<td align="left">Betula_alnoides</td>
</tr>
<tr class="even">
<td align="left">SAMPLE10</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU1</td>
<td align="right">372</td>
<td align="left">MF171078</td>
<td align="right">98.71</td>
<td align="left">Pentaphylacaceae</td>
<td align="left">Pentaphylax_euryoides</td>
</tr>
<tr class="odd">
<td align="left">SAMPLE10</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU2</td>
<td align="right">14</td>
<td align="left">KY214931</td>
<td align="right">98.33</td>
<td align="left">Musaceae</td>
<td align="left">Musa_ornata</td>
</tr>
<tr class="even">
<td align="left">SAMPLE10</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU2</td>
<td align="right">14</td>
<td align="left">KY214926</td>
<td align="right">96.23</td>
<td align="left">Musaceae</td>
<td align="left">Musa_sp.</td>
</tr>
<tr class="odd">
<td align="left">SAMPLE10</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU2</td>
<td align="right">14</td>
<td align="left">KY214930</td>
<td align="right">95.73</td>
<td align="left">Musaceae</td>
<td align="left">Musa_basjoo</td>
</tr>
<tr class="even">
<td align="left">SAMPLE10</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU2</td>
<td align="right">14</td>
<td align="left">KY214927</td>
<td align="right">97.67</td>
<td align="left">Musaceae</td>
<td align="left">Musella_lasiocarpa</td>
</tr>
<tr class="odd">
<td align="left">SAMPLE10</td>
<td align="left">ITS1</td>
<td align="left">R2</td>
<td align="left">OTU1</td>
<td align="right">357</td>
<td align="left">AM233397</td>
<td align="right">95.62</td>
<td align="left">Apocynaceae</td>
<td align="left">Secamone_filiformis</td>
</tr>
<tr class="even">
<td align="left">SAMPLE10</td>
<td align="left">ITS1</td>
<td align="left">R2</td>
<td align="left">OTU1</td>
<td align="right">357</td>
<td align="left">FR832858</td>
<td align="right">96.69</td>
<td align="left">Rubiaceae</td>
<td align="left">Tricalysia_perrieri</td>
</tr>
</tbody>
</table>
<p>As you see, the following information are included in the data:</p>
<ul>
<li><code>sample</code> is simply the name of the sample,</li>
<li><code>its</code> is either ITS1 or ITS2 and tells which set of PCR primers has been used,</li>
<li><code>replicate</code> contains information on which replicate the sequences come from,</li>
<li><code>OTU</code> is a unique identifier of the so-called Operational Taxonomic Unit, an OTU often corresponds to one species but not always. Sometimes 2 OTUs represent the same species, sometimes 1 OTU consists of more than one species,</li>
<li><code>size</code> is the number of reads that support that particular OTU,</li>
<li><code>hit</code> is the BLAST hit identifier. The 4 top BLAST hits are reported per OTU,</li>
<li><code>perc_identity</code> is the percentage identity of the sequence to the BLAST hit,</li>
<li><code>family</code> is the identified plant family,</li>
<li><code>species</code> is the identified plant species.</li>
</ul>
</div>
<div id="number-of-replicates-per-species" class="section level2">
<h2><span class="header-section-number">6.4</span> Number of replicates per species</h2>
<p>Create a new dataset <code>species</code> that contains an extra column <code>n_replicates</code>. The column contains number of replicates this particular species is present in. Do it per sample and its.</p>
<pre class="r"><code>species &lt;- species_orig %&gt;%
  group_by(sample, its, species) %&gt;%
  mutate(n_replicates = n_distinct(replicate)) %&gt;%
  ungroup()</code></pre>
<table>
<thead>
<tr class="header">
<th align="left">sample</th>
<th align="left">its</th>
<th align="left">replicate</th>
<th align="left">OTU</th>
<th align="right">size</th>
<th align="left">hit</th>
<th align="right">perc_ident</th>
<th align="left">family</th>
<th align="left">species</th>
<th align="right">n_replicates</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">SAMPLE10</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU1</td>
<td align="right">372</td>
<td align="left">KX522674</td>
<td align="right">97.08</td>
<td align="left">Anacardiaceae</td>
<td align="left">Spondias_tuberosa</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td align="left">SAMPLE10</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU1</td>
<td align="right">372</td>
<td align="left">AJ783644</td>
<td align="right">97.53</td>
<td align="left">Betulaceae</td>
<td align="left">Betula_populifolia</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">SAMPLE10</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU1</td>
<td align="right">372</td>
<td align="left">AJ783641</td>
<td align="right">97.53</td>
<td align="left">Betulaceae</td>
<td align="left">Betula_alnoides</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">SAMPLE10</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU1</td>
<td align="right">372</td>
<td align="left">MF171078</td>
<td align="right">98.71</td>
<td align="left">Pentaphylacaceae</td>
<td align="left">Pentaphylax_euryoides</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">SAMPLE10</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU2</td>
<td align="right">14</td>
<td align="left">KY214931</td>
<td align="right">98.33</td>
<td align="left">Musaceae</td>
<td align="left">Musa_ornata</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">SAMPLE10</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU2</td>
<td align="right">14</td>
<td align="left">KY214926</td>
<td align="right">96.23</td>
<td align="left">Musaceae</td>
<td align="left">Musa_sp.</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">SAMPLE10</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU2</td>
<td align="right">14</td>
<td align="left">KY214930</td>
<td align="right">95.73</td>
<td align="left">Musaceae</td>
<td align="left">Musa_basjoo</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">SAMPLE10</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU2</td>
<td align="right">14</td>
<td align="left">KY214927</td>
<td align="right">97.67</td>
<td align="left">Musaceae</td>
<td align="left">Musella_lasiocarpa</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">SAMPLE10</td>
<td align="left">ITS1</td>
<td align="left">R2</td>
<td align="left">OTU1</td>
<td align="right">357</td>
<td align="left">AM233397</td>
<td align="right">95.62</td>
<td align="left">Apocynaceae</td>
<td align="left">Secamone_filiformis</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">SAMPLE10</td>
<td align="left">ITS1</td>
<td align="left">R2</td>
<td align="left">OTU1</td>
<td align="right">357</td>
<td align="left">FR832858</td>
<td align="right">96.69</td>
<td align="left">Rubiaceae</td>
<td align="left">Tricalysia_perrieri</td>
<td align="right">1</td>
</tr>
</tbody>
</table>
</div>
<div id="filter-out-unwanted-samples" class="section level2">
<h2><span class="header-section-number">6.5</span> Filter out unwanted samples</h2>
<p>Now, your task is to filter out all but your <code>sample_name</code> samples from the dataset.
Call the resulting dataset <code>my_sample</code>.</p>
<pre class="r"><code>my_sample &lt;- species %&gt;%
  filter(sample == sample_name)</code></pre>
</div>
<div id="missing-observations" class="section level2">
<h2><span class="header-section-number">6.6</span> Missing observations</h2>
<p>What happens if a set of primers failed to amplify or if one replicate was lost?
Use <code>complete()</code> to make sure you have <code>NA</code> values in such cases.</p>
<pre class="r"><code>my_sample &lt;- my_sample %&gt;%
  complete(its = c(&quot;ITS1&quot;, &quot;ITS2&quot;),
           replicate = c(&quot;R1&quot;,&quot;R2&quot;,&quot;R3&quot;)
           )</code></pre>
</div>
<div id="sorting-issue" class="section level2">
<h2><span class="header-section-number">6.7</span> Sorting issue</h2>
<p>Look, the first sample in the table is <code>SAMPLE10</code>. Why not <code>SAMPLE1</code>?
Thats a sorting issue: if sorted as character, 10 will come before 1. WE have to fix this by adding trailing zero to the values in <code>OTU</code>. We do not expect more than 99 OTUs in a sample, so it is ok with only one trailing 0 (otherwise the 100-th sample will spoil our sorting and come out like: SAMPLE100, SAMPLE01, SAMPLE10). We will need to use regular expression:</p>
<ul>
<li>all values in the <code>OTU</code> column that follow pattern “OTU<em>digit</em>” we need to change to “OTU0<em>digit</em>”. Regular expression that matches this is <code>OTU([0-9]$)</code> and it should be replaced by: <code>OTU0\\1</code>. Ask your TAs to explain this if you do not know much about regular expressions and pattern matching.</li>
</ul>
<pre class="r"><code>my_sample &lt;- my_sample %&gt;% mutate(OTU = str_replace(OTU,
                                  pattern = &quot;OTU([0-9]$)&quot;,
                                  replacement = &quot;OTU0\\1&quot;
                                )
                          )</code></pre>
<table>
<thead>
<tr class="header">
<th align="left">its</th>
<th align="left">replicate</th>
<th align="left">sample</th>
<th align="left">OTU</th>
<th align="right">size</th>
<th align="left">hit</th>
<th align="right">perc_ident</th>
<th align="left">family</th>
<th align="left">species</th>
<th align="right">n_replicates</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">SAMPLE12</td>
<td align="left">OTU01</td>
<td align="right">4169</td>
<td align="left">LC089998</td>
<td align="right">99.44</td>
<td align="left">Brassicaceae</td>
<td align="left">Capsella_bursa-pastoris</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">SAMPLE12</td>
<td align="left">OTU02</td>
<td align="right">2686</td>
<td align="left">KP794392</td>
<td align="right">98.19</td>
<td align="left">Euphorbiaceae</td>
<td align="left">Tragia_cf.</td>
<td align="right">3</td>
</tr>
<tr class="odd">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">SAMPLE12</td>
<td align="left">OTU02</td>
<td align="right">2686</td>
<td align="left">KP794390</td>
<td align="right">98.19</td>
<td align="left">Euphorbiaceae</td>
<td align="left">Tragia_chlorocaulon</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">SAMPLE12</td>
<td align="left">OTU02</td>
<td align="right">2686</td>
<td align="left">KP794378</td>
<td align="right">98.19</td>
<td align="left">Euphorbiaceae</td>
<td align="left">Tragia_bahiensis</td>
<td align="right">3</td>
</tr>
<tr class="odd">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">SAMPLE12</td>
<td align="left">OTU02</td>
<td align="right">2686</td>
<td align="left">KP794320</td>
<td align="right">98.19</td>
<td align="left">Euphorbiaceae</td>
<td align="left">Bia_alienata</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">SAMPLE12</td>
<td align="left">OTU04</td>
<td align="right">1353</td>
<td align="left">GQ396671</td>
<td align="right">97.44</td>
<td align="left">Euphorbiaceae</td>
<td align="left">Triadica_sebifera</td>
<td align="right">3</td>
</tr>
<tr class="odd">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">SAMPLE12</td>
<td align="left">OTU04</td>
<td align="right">1353</td>
<td align="left">KX522674</td>
<td align="right">96.25</td>
<td align="left">Anacardiaceae</td>
<td align="left">Spondias_tuberosa</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">SAMPLE12</td>
<td align="left">OTU04</td>
<td align="right">1353</td>
<td align="left">KY860928</td>
<td align="right">96.23</td>
<td align="left">Adoxaceae</td>
<td align="left">Viburnum_prunifolium</td>
<td align="right">3</td>
</tr>
<tr class="odd">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">SAMPLE12</td>
<td align="left">OTU04</td>
<td align="right">1353</td>
<td align="left">KX757310</td>
<td align="right">96.23</td>
<td align="left">Caryophyllaceae</td>
<td align="left">Silene_caesia</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">SAMPLE12</td>
<td align="left">OTU06</td>
<td align="right">577</td>
<td align="left">KY860926</td>
<td align="right">100.00</td>
<td align="left">Asteraceae</td>
<td align="left">Taraxacum_officinale</td>
<td align="right">3</td>
</tr>
</tbody>
</table>
</div>
<div id="supporting-reads" class="section level2">
<h2><span class="header-section-number">6.8</span> Supporting reads</h2>
<p>Sometimes, an OTU generates two or more top BLAST hits that come from the same species. We have decided to sum reads in such cases. Do it!</p>
<pre class="r"><code>  my_sample &lt;- my_sample %&gt;%
    ungroup() %&gt;%
    group_by(sample, its, replicate, OTU, species, n_replicates) %&gt;%
    summarise(n_reads = sum(size)) %&gt;%
    ungroup() %&gt;%
    group_by(its, species, OTU)</code></pre>
<table>
<thead>
<tr class="header">
<th align="left">sample</th>
<th align="left">its</th>
<th align="left">replicate</th>
<th align="left">OTU</th>
<th align="left">species</th>
<th align="right">n_replicates</th>
<th align="right">n_reads</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">NA</td>
<td align="left">ITS2</td>
<td align="left">R1</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="even">
<td align="left">NA</td>
<td align="left">ITS2</td>
<td align="left">R2</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="odd">
<td align="left">NA</td>
<td align="left">ITS2</td>
<td align="left">R3</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="even">
<td align="left">SAMPLE12</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU01</td>
<td align="left">Capsella_bursa-pastoris</td>
<td align="right">3</td>
<td align="right">4169</td>
</tr>
<tr class="odd">
<td align="left">SAMPLE12</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU02</td>
<td align="left">Bia_alienata</td>
<td align="right">3</td>
<td align="right">2686</td>
</tr>
<tr class="even">
<td align="left">SAMPLE12</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU02</td>
<td align="left">Tragia_bahiensis</td>
<td align="right">3</td>
<td align="right">2686</td>
</tr>
<tr class="odd">
<td align="left">SAMPLE12</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU02</td>
<td align="left">Tragia_cf.</td>
<td align="right">3</td>
<td align="right">2686</td>
</tr>
<tr class="even">
<td align="left">SAMPLE12</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU02</td>
<td align="left">Tragia_chlorocaulon</td>
<td align="right">3</td>
<td align="right">2686</td>
</tr>
<tr class="odd">
<td align="left">SAMPLE12</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU04</td>
<td align="left">Silene_caesia</td>
<td align="right">3</td>
<td align="right">1353</td>
</tr>
<tr class="even">
<td align="left">SAMPLE12</td>
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU04</td>
<td align="left">Spondias_tuberosa</td>
<td align="right">3</td>
<td align="right">1353</td>
</tr>
</tbody>
</table>
</div>
<div id="within-otu-species-diversity" class="section level2">
<h2><span class="header-section-number">6.9</span> Within-OTU species diversity</h2>
<p>Now, we want to see how many identifications an OTU got. Implement this. Store the result in a new tibble <code>diversity</code>.</p>
<pre class="r"><code>diversity &lt;- my_sample %&gt;%
  ungroup() %&gt;%
  group_by(its, replicate, OTU) %&gt;%
  summarise(n_species = n())</code></pre>
<table>
<thead>
<tr class="header">
<th align="left">its</th>
<th align="left">replicate</th>
<th align="left">OTU</th>
<th align="right">n_species</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU01</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU02</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU04</td>
<td align="right">4</td>
</tr>
<tr class="even">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU06</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU07</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU08</td>
<td align="right">2</td>
</tr>
<tr class="odd">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU10</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU11</td>
<td align="right">3</td>
</tr>
<tr class="odd">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU13</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU15</td>
<td align="right">2</td>
</tr>
</tbody>
</table>
</div>
<div id="adding-diversity-data" class="section level2">
<h2><span class="header-section-number">6.10</span> Adding diversity data</h2>
<p>Add the <code>diversity</code> data to <code>my_sample</code> using appropriate <code>join</code> function.
Also, remove the column with sample names since we are dealing with only one sample.</p>
<pre class="r"><code>my_sample &lt;- my_sample %&gt;%
  left_join(diversity) %&gt;%
  select(-sample)</code></pre>
<table>
<thead>
<tr class="header">
<th align="left">its</th>
<th align="left">replicate</th>
<th align="left">OTU</th>
<th align="left">species</th>
<th align="right">n_replicates</th>
<th align="right">n_reads</th>
<th align="right">n_species</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ITS2</td>
<td align="left">R1</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">ITS2</td>
<td align="left">R2</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">ITS2</td>
<td align="left">R3</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU01</td>
<td align="left">Capsella_bursa-pastoris</td>
<td align="right">3</td>
<td align="right">4169</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU02</td>
<td align="left">Bia_alienata</td>
<td align="right">3</td>
<td align="right">2686</td>
<td align="right">4</td>
</tr>
<tr class="even">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU02</td>
<td align="left">Tragia_bahiensis</td>
<td align="right">3</td>
<td align="right">2686</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU02</td>
<td align="left">Tragia_cf.</td>
<td align="right">3</td>
<td align="right">2686</td>
<td align="right">4</td>
</tr>
<tr class="even">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU02</td>
<td align="left">Tragia_chlorocaulon</td>
<td align="right">3</td>
<td align="right">2686</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU04</td>
<td align="left">Silene_caesia</td>
<td align="right">3</td>
<td align="right">1353</td>
<td align="right">4</td>
</tr>
<tr class="even">
<td align="left">ITS1</td>
<td align="left">R1</td>
<td align="left">OTU04</td>
<td align="left">Spondias_tuberosa</td>
<td align="right">3</td>
<td align="right">1353</td>
<td align="right">4</td>
</tr>
</tbody>
</table>
<ul>
<li>what columns did we join on in our example solution?</li>
</ul>
</div>
<div id="visualising-the-data" class="section level2">
<h2><span class="header-section-number">6.11</span> Visualising the data</h2>
<p>Can you think of a good way of visualising the data? Think of:</p>
<ul>
<li>Type of plot you want. The simpler the better?</li>
<li>Which variables you would like to visualise? We have chosen <code>ITS</code>, <code>replicate</code>, <code>n_reads</code>, <code>n_replicates</code>, <code>OTU</code>, <code>threshold</code>, <code>n_species</code> and <code>specie</code>. Well, pretty much all of them :-)</li>
<li>How do you represent variables: colours, shapes, separate plots, sizes?</li>
<li>What transformations do you need to apply before visualising the data?</li>
</ul>
<p>Our example solution involves some <code>ggplot2</code> magics. But would base-R be good enough for this type of plot?</p>
<div class="figure" style="text-align: left"><span id="fig:unnamed-chunk-54"></span>
<img src="lab_files/figure-html/unnamed-chunk-54-1.svg" alt="Red -- only one species in the current OTU, blue -- identification above the threshold, grey -- identification below the threshold" width="864" />
<p class="caption">
Figure 6.1: Red – only one species in the current OTU, blue – identification above the threshold, grey – identification below the threshold
</p>
</div>
</div>
</div>
<div id="wildlife-aircraft-strikes-challenge" class="section level1">
<h1><span class="header-section-number">7</span> Wildlife Aircraft Strikes Challenge</h1>
<p>Use the <a href="https://www.kaggle.com/faa/wildlife-strikes">FAA report</a> and <code>tidyverse</code> to learn more about aircraft incidents with wildlife. Use your imagination and <a href="https://nycdatascience.com/blog/student-works/exploring-aviation-accidents-from-1908-through-the-present/">NYC data science blog</a> for inspiration!</p>
</div>
<div id="session-info" class="section level1">
<h1><span class="header-section-number">8</span> Session Info</h1>
<ul>
<li>This document has been created in RStudio using R Markdown and related packages.</li>
<li>For R Markdown, see <a href="http://rmarkdown.rstudio.com" class="uri">http://rmarkdown.rstudio.com</a></li>
<li>For details about the OS, packages and versions, see detailed information below:</li>
</ul>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 3.5.0 (2018-04-23)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS  10.14.4
## 
## Matrix products: default
## BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] lubridate_1.7.4             ShortRead_1.40.0           
##  [3] GenomicAlignments_1.18.0    SummarizedExperiment_1.12.0
##  [5] DelayedArray_0.8.0          matrixStats_0.54.0         
##  [7] Biobase_2.42.0              Rsamtools_1.34.0           
##  [9] GenomicRanges_1.34.0        GenomeInfoDb_1.18.0        
## [11] Biostrings_2.50.1           XVector_0.22.0             
## [13] IRanges_2.16.0              S4Vectors_0.20.0           
## [15] BiocParallel_1.16.0         BiocGenerics_0.28.0        
## [17] here_0.1                    eulerr_5.1.0               
## [19] bsplus_0.1.1                forcats_0.4.0              
## [21] stringr_1.4.0               dplyr_0.8.1                
## [23] purrr_0.3.2                 readr_1.3.1                
## [25] tidyr_0.8.3                 tibble_2.1.1               
## [27] tidyverse_1.2.1             ggplot2_3.1.1              
## [29] captioner_2.2.3             bookdown_0.11              
## [31] knitr_1.23                 
## 
## loaded via a namespace (and not attached):
##  [1] httr_1.4.0             jsonlite_1.6           modelr_0.1.4          
##  [4] assertthat_0.2.1       highr_0.8              latticeExtra_0.6-28   
##  [7] GenomeInfoDbData_1.2.0 cellranger_1.1.0       yaml_2.2.0            
## [10] pillar_1.4.0           backports_1.1.4        lattice_0.20-38       
## [13] glue_1.3.1             digest_0.6.19          RColorBrewer_1.1-2    
## [16] polyclip_1.10-0        rvest_0.3.4            colorspace_1.4-1      
## [19] htmltools_0.3.6        Matrix_1.2-17          plyr_1.8.4            
## [22] pkgconfig_2.0.2        broom_0.5.2            haven_2.1.0           
## [25] zlibbioc_1.28.0        scales_1.0.0           generics_0.0.2        
## [28] withr_2.1.2            lazyeval_0.2.2         cli_1.1.0             
## [31] magrittr_1.5           crayon_1.3.4           readxl_1.3.1          
## [34] evaluate_0.14          nlme_3.1-140           hwriter_1.3.2         
## [37] xml2_1.2.0             tools_3.5.0            hms_0.4.2             
## [40] munsell_0.5.0          compiler_3.5.0         rlang_0.3.4           
## [43] grid_3.5.0             RCurl_1.95-4.12        rstudioapi_0.10       
## [46] labeling_0.3           bitops_1.0-6           rmarkdown_1.13.1      
## [49] gtable_0.3.0           R6_2.4.0               rprojroot_1.3-2       
## [52] polylabelr_0.1.0       stringi_1.4.3          Rcpp_1.0.1            
## [55] tidyselect_0.2.5       xfun_0.7</code></pre>
<p style="text-align: left; font-size: small;">
Page built on: <i class="fa fa-calendar" aria-hidden="true"></i> 12-Jun-2019 at <i class="fa fa-clock-o" aria-hidden="true"></i> 12:04:16.
</p>
<hr/>
<div style="padding-bottom: 1.5em">
<p><span style="float:left; vertical-align:middle">
<b>2018</b> | <a href="https://www.scilifelab.se/">SciLifeLab</a> &gt; <a href="https://nbis.se/">NBIS</a> &gt; <a href="https://nbisweden.github.io/workshop-RaukR-1806/">RaukR</a>
</span>
<span style="float:right; vertical-align:middle">
<span class="footericon" style="padding-right:4px; padding-left:4px">
<a href="https://nbisweden.github.io/workshop-RaukR-1806/"><img src="assets/icons8-globe-26.png" alt="website" border="0" style="height:15px"></a>
</span>
<span class="footericon" style="padding-right:4px; padding-left:4px">
<a href="https://twitter.com/hashtag/RaukR?src=hash"><img src="assets/icons8-twitter-26.png" alt="twitter" border="0" style="height:15px"></a>
</span>
</span></p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
