<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
  <head>
    <title>Tidy Work in Tidyverse</title>
    <meta charset="utf-8" />
    <meta name="author" content="Marcin Kierczak" />
    <meta name="keywords" content="r, RaukR, markdown" />
    <link href="libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
    <link href="libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
    <link id="font-awesome-1-attachment" rel="attachment" href="libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>
    <script src="libs/kePrint-0.0.1/kePrint.js"></script>
    <link rel="stylesheet" href="assets/presentation.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Tidy Work in Tidyverse
## RaukR 2019 • Advanced R for Bioinformatics
### <b>Marcin Kierczak</b>
### NBIS, SciLifeLab

---

exclude: true
count: false


&lt;link href="https://fonts.googleapis.com/css?family=Roboto|Source+Sans+Pro:300,400,600|Ubuntu+Mono&amp;amp;subset=latin-ext" rel="stylesheet"&gt;





&lt;!-- ----------------- Only edit title &amp; author above this ----------------- --&gt;





## Tidyverse -- What is it all About?

* [Tidyverse](http://www.tidyverse.org) is a collection of packages. 
* Created by [Hadley Wickham](http://hadley.nz).
* Gains popularity, on the way to become a *de facto* standard in data analyses.
* Knowing how to use it can increase your salary :-)
* A philosophy of programming or a programing paradigm.
* Everything is about the flow of *tidy data*.
.center[
&lt;img src="assets_tidyverse/hex-tidyverse.png", style="height:200px;"&gt;
&lt;img src="assets_tidyverse/Hadley-wickham2016-02-04.jpeg", style="height:200px;"&gt;
&lt;img src="assets_tidyverse/RforDataScience.jpeg", style="height:200px;"&gt;
]
.vsmall[sources of images: www.tidyverse.org, Wikipedia, www.tidyverse.org]

---
name: tidyverse_workflow

## Typical Tidyverse Workflow
The tidyverse curse?&lt;br&gt;&lt;br&gt;
--
*Navigating the balance between base R and the tidyverse is a challenge to learn.*
.right[.small[-- [Robert A. Muenchen](http://r4stats.com/articles/why-r-is-hard-to-learn/)]]
&lt;br&gt;&lt;br&gt;
--
&lt;img src="assets_tidyverse/tidyverse-flow.png", style="height:400px;"&gt;&lt;br&gt;
.vsmall[source: http://www.storybench.org/getting-started-with-tidyverse-in-r/]  

---
name: intro_to_pipes
## Introduction to Pipes
.pull-left-50[
  .center[
    &lt;img src="assets_tidyverse/MagrittePipe.jpg" width="300" style="display: block; margin: auto auto auto 0;" /&gt;
  ]
  .vsmall[
    Rene Magritt, *La trahison des images*, [Wikimedia Commons](https://en.wikipedia.org/wiki/The_Treachery_of_Images#/media/File:MagrittePipe.jpg)
  ]
  &lt;br&gt;&amp;nbsp;
  .center[
    &lt;img src="assets_tidyverse/magrittr.png" width="150" style="display: block; margin: auto auto auto 0;" /&gt;
  ]
]
--
.pull-right-50[
* Let the data flow.
* *Ceci n'est pas une pipe* -- `magrittr`
* The `%&gt;%` pipe:
  + `x %&gt;% f` `\(\equiv\)` `f(x)`
  + `x %&gt;% f(y)` `\(\equiv\)` `f(x, y)`
  + `x %&gt;% f %&gt;% g %&gt;% h` `\(\equiv\)` `h(g(f(x)))`
]
--
.pull-right-50[
instead of writing this:

```r
data &lt;- iris
data &lt;- head(data, n=3)
```
]
--
.pull-right-50[
write this:

```r
iris %&gt;% head(n=3)
```

```
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 1          5.1         3.5          1.4         0.2  setosa
## 2          4.9         3.0          1.4         0.2  setosa
## 3          4.7         3.2          1.3         0.2  setosa
```
]

---
name: other_pipes_T
## Other Types of `magrittr` Pipes -- `%T&gt;%`

.pull-left-50[
The %T&gt;% pipe is useful when you call a function for its side effects:

```r
rnorm(50) %&gt;%
  matrix(ncol = 2) %&gt;%
  plot() %&gt;%
  summary()
```

```
## Length  Class   Mode 
##      0   NULL   NULL
```

&lt;img src="tidyverse_presentation_files/figure-html/magrittr2a-1.png" style="display: block; margin: auto auto auto 0;" /&gt;
]
--
.pull-right-50[

```r
rnorm(50) %&gt;%
  matrix(ncol = 2) %T&gt;%
  plot() %&gt;%
  summary()
```

```
##        V1                V2          
##  Min.   :-1.8005   Min.   :-1.63074  
##  1st Qu.:-0.7070   1st Qu.:-0.72419  
##  Median :-0.1815   Median : 0.21506  
##  Mean   :-0.1333   Mean   :-0.08634  
##  3rd Qu.: 0.3385   3rd Qu.: 0.46185  
##  Max.   : 1.8652   Max.   : 0.89255
```

&lt;img src="tidyverse_presentation_files/figure-html/magrittr2b-1.png" style="display: block; margin: auto auto auto 0;" /&gt;
]

---
name: the_splitting_pipe

## Other Types of `magrittr` Pipes -- `%$%`

```r
iris %&gt;% cor(Sepal.Length, Sepal.Width)
```

```
## Error in pmatch(use, c("all.obs", "complete.obs", "pairwise.complete.obs", : object 'Sepal.Width' not found
```


We need the `%$%` pipe with exposition of variables:


```r
iris %$% cor(Sepal.Length, Sepal.Width)
```

```
## [1] -0.1175698
```
This is because the `cor` function does not have the `data` argument (which also should be the first argument of a pipe-friendly function).

### The %&lt;&gt;% Pipe
It exists but can lead to somewhat confusing code.  
`x %&lt;&gt;% f` `\(\equiv\)` `x &lt;- f(x)`

```r
M &lt;- matrix(rnorm(16), nrow=4)
M %&lt;&gt;% colSums()
M
```

```
## [1] -1.8570306 -0.1943518 -0.8976997  3.7219083
```
---
name: magrittr_placeholder

## Placeholders in `magrittr` Pipes
Sometimes we want to pass the resulting data to *other than the first* argument of the next function in chain. `magritter` provides placeholder mechanism for this:
* `x %&gt;% f(y, .)` `\(\equiv\)` `f(y, x)`,
* `x %&gt;% f(y, z = .)` `\(\equiv\)` `f(y, z = x)`.

But for nested expressions:
* `x %&gt;% f(a = p(.), b = q(.))` `\(\equiv\)` `f(x, a = p(x), b = q(x))`,
* `x %&gt;% {f(a = p(.), b = q(.))}` `\(\equiv\)` `f(a = p(x), b = q(x))`.  

Examples:

```r
M &lt;- rnorm(4) %&gt;% matrix(nrow = 2)
M %&gt;% `%*%`(., .)
```

```
##           [,1]      [,2]
## [1,]  1.261849 0.4939613
## [2,] -1.223489 1.1420257
```


```r
print_M_summ &lt;- function(nrow, ncol) {
  paste0('Matrix M has: ', nrow, ' rows and ', ncol, ' columns.')
}
M %&gt;% {print_M_summ(nrow(.), ncol(.))}
```

```
## [1] "Matrix M has: 2 rows and 2 columns."
```

---
name: tibble_intro

## Tibbles
.pull-left-50[
  &lt;img src="assets_tidyverse/hex-tibble.png" width="160" style="display: block; margin: auto;" /&gt;
  
  ```r
  as_tibble(iris)
  ```
  
  ```
  ## # A tibble: 150 x 5
  ##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
  ##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;fct&gt;  
  ##  1          5.1         3.5          1.4         0.2 setosa 
  ##  2          4.9         3            1.4         0.2 setosa 
  ##  3          4.7         3.2          1.3         0.2 setosa 
  ##  4          4.6         3.1          1.5         0.2 setosa 
  ##  5          5           3.6          1.4         0.2 setosa 
  ##  6          5.4         3.9          1.7         0.4 setosa 
  ##  7          4.6         3.4          1.4         0.3 setosa 
  ##  8          5           3.4          1.5         0.2 setosa 
  ##  9          4.4         2.9          1.4         0.2 setosa 
  ## 10          4.9         3.1          1.5         0.1 setosa 
  ## # … with 140 more rows
  ```
]

.pull-right-50[
* `tibble` is one of the unifying features of tidyverse,
* it is a *better* `data.frame` realization,
* objects `data.frame` can be coerced to `tibble` using `as_tibble()`


```r
  tibble(
    x = 1,          # recycling
    y = runif(50), 
    z = x + y^2,
    outcome = rnorm(50)
  )
```

```
## # A tibble: 50 x 4
##        x      y     z  outcome
##    &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;
##  1     1 0.528   1.28  0.00203
##  2     1 0.626   1.39 -3.18   
##  3     1 0.930   1.86 -1.78   
##  4     1 0.702   1.49 -0.199  
##  5     1 0.0807  1.01  0.129  
##  6     1 0.644   1.41 -2.12   
##  7     1 0.0435  1.00 -1.61   
##  8     1 0.880   1.77  0.451  
##  9     1 0.250   1.06  1.50   
## 10     1 0.0855  1.01  0.568  
## # … with 40 more rows
```
]

---
name: tibble2

## More on Tibbles

* When you print a `tibble`:
  + all columns that fit the screen are shown,
  + first 10 rows are shown,
  + data type for each column is shown.


```r
as_tibble(cars)
```

```
## # A tibble: 50 x 2
##    speed  dist
##    &lt;dbl&gt; &lt;dbl&gt;
##  1     4     2
##  2     4    10
##  3     7     4
##  4     7    22
##  5     8    16
##  6     9    10
##  7    10    18
##  8    10    26
##  9    10    34
## 10    11    17
## # … with 40 more rows
```

* `my_tibble %&gt;% print(n = 50, width = Inf)`,
* `options(tibble.print_min = 15, tibble.print_max = 25)`,
* `options(dplyr.print_min = Inf)`,
* `options(tibble.width = Inf)`

---
name: tibble2

## Subsetting Tibbles


```r
vehicles &lt;- as_tibble(cars[1:5,])

vehicles[['speed']]
vehicles[[1]]
vehicles$speed

# Using placeholders

vehicles %&gt;% .$dist
vehicles %&gt;% .[['dist']]
vehicles %&gt;% .[[2]]
```

```
## [1] 4 4 7 7 8
## [1] 4 4 7 7 8
## [1] 4 4 7 7 8
## [1]  2 10  4 22 16
## [1]  2 10  4 22 16
## [1]  2 10  4 22 16
```
--
**Note!** Not all old R functions work with tibbles, than you have to use `as.data.frame(my_tibble)`.

---
name: tibbles_partial_matching

## Tibbles are Stricter than `data.frames`


```r
cars$spe      # partial matching
```

```
## [1] 4 4 7 7 8
```

```r
vehicles$spe  # no partial matching
```

```
## Warning: Unknown or uninitialised column: 'spe'.
```

```
## NULL
```

```r
cars$gear
```

```
## NULL
```

```r
vehicles$gear
```

```
## Warning: Unknown or uninitialised column: 'gear'.
```

```
## NULL
```

---
name: loading_data

## Loading Data
In `tidyverse` you import data using `readr` package that provides a number of useful data import functions:
* `read_delim()` a generic function for reading *-delimited files. There are a number of convenience wrappers:
  + `read_csv()` used to read comma-delimited files,
  + `read_csv2()` reads semicolon-delimited files, 
  `read_tsv()` that reads tab-delimited files.
* `read_fwf` for reading fixed-width files with its wrappers:
  + fwf_widths() for width-based reading,
  + fwf_positions() for positions-based reading and
  + read_table() for reading white space-delimited fixed-width files.
* `read_log()` for reading Apache-style logs.

The most commonly used `read_csv()` has some familiar arguments like:
* `skip` -- to specify the number of rows to skip (headers),
* `col_names` -- to supply a vector of column names,
* `comment` -- to specify what character designates a comment,
* `na` -- to specify how missing values are represented.

---
name: parse_functions

## Under the Hood -- `parse_*` Functions 
Under the hood, data-reading functions use `parse_*` functions:


```r
parse_double("42.24")
```

```
## [1] 42.24
```


```r
parse_number("272'555'849,55", 
             locale = locale(decimal_mark = ",", 
                             grouping_mark = "'"
                            )
             )
```

```
## [1] 272555850
```


```r
parse_number(c('100%', 'price: 500$', '21sek', '42F'))
```

```
## [1] 100 500  21  42
```

---
name: parsing_strings

## Parsing Strings

* Strings can be represented in different encodings:

```r
text1 &lt;- 'På en ö är en å'
text2 &lt;- 'Zażółć gęślą jaźń'
```

```r
text1
charToRaw(text2)
parse_character(text1, locale = locale(encoding = 'UTF-8'))
guess_encoding(charToRaw("Test"))
guess_encoding(charToRaw(text1))
```

```
## [1] "På en ö är en å"
##  [1] 5a 61 c5 bc c3 b3 c5 82 c4 87 20 67 c4 99 c5 9b 6c c4 85 20 6a 61 c5
## [24] ba c5 84
## [1] "På en ö är en å"
## # A tibble: 1 x 2
##   encoding confidence
##   &lt;chr&gt;         &lt;dbl&gt;
## 1 ASCII             1
## # A tibble: 4 x 2
##   encoding   confidence
##   &lt;chr&gt;           &lt;dbl&gt;
## 1 UTF-8            1   
## 2 ISO-8859-1       0.83
## 3 ISO-8859-9       0.33
## 4 ISO-8859-2       0.3
```

---
name: parsing_factors

## Parsing Factors

* R is using factors to represent cathegorical variables. 
* Supply known levels to `parse_factor` so that it warns you when an unknown level is present in the data:

```r
landscapes &lt;- c('mountains', 'swamps', 'seaside')
parse_factor(c('mountains', 'plains', 'seaside', 'swamps'), 
             levels = landscapes)
```

```
## Warning: 1 parsing failure.
## row col           expected actual
##   2  -- value in level set plains
```

```
## [1] mountains &lt;NA&gt;      seaside   swamps   
## attr(,"problems")
## # A tibble: 1 x 4
##     row   col expected           actual
##   &lt;int&gt; &lt;int&gt; &lt;chr&gt;              &lt;chr&gt; 
## 1     2    NA value in level set plains
## Levels: mountains swamps seaside
```
---
name: parsing_other_functions

## Other Parsing Functions

`parse_`
* `vector`, `time`, `number`, `logical`, `integer`, `double`, `character`, `date`, `datetime`,
* `guess`


```r
guess_parser("2018-06-11 09:00:00")
parse_guess("2018-06-11 09:00:00")

guess_parser(c(1, 2.3, "23$", "54%"))
parse_guess(c(1, 2.3, "23$", "54%"))
```

```
## [1] "datetime"
## [1] "2018-06-11 09:00:00 UTC"
## [1] "character"
## [1] "1"   "2.3" "23$" "54%"
```

---
name: readr

## Importing Data Using `readr`

When reading and parsing a file, `readr` attempts to guess proper parser for each column by looking at the 1000 first rows.

```r
tricky_dataset &lt;- read_csv(readr_example('challenge.csv'))
```

```
## Parsed with column specification:
## cols(
##   x = col_double(),
##   y = col_logical()
## )
```

```
## Warning: 1000 parsing failures.
##  row col           expected     actual                                                                                         file
## 1001   y 1/0/T/F/TRUE/FALSE 2015-01-16 '/Library/Frameworks/R.framework/Versions/3.5/Resources/library/readr/extdata/challenge.csv'
## 1002   y 1/0/T/F/TRUE/FALSE 2018-05-18 '/Library/Frameworks/R.framework/Versions/3.5/Resources/library/readr/extdata/challenge.csv'
## 1003   y 1/0/T/F/TRUE/FALSE 2015-09-05 '/Library/Frameworks/R.framework/Versions/3.5/Resources/library/readr/extdata/challenge.csv'
## 1004   y 1/0/T/F/TRUE/FALSE 2012-11-28 '/Library/Frameworks/R.framework/Versions/3.5/Resources/library/readr/extdata/challenge.csv'
## 1005   y 1/0/T/F/TRUE/FALSE 2020-01-13 '/Library/Frameworks/R.framework/Versions/3.5/Resources/library/readr/extdata/challenge.csv'
## .... ... .................. .......... ............................................................................................
## See problems(...) for more details.
```
OK, so there are some parsing failures. We can examine them more closely using `problems()` as suggested in the above output.

---
name: readr_problems
## Looking at Problematic Columns


```r
p &lt;- problems(tricky_dataset)
p
```

```
## # A tibble: 1,000 x 5
##      row col   expected      actual   file                                 
##    &lt;int&gt; &lt;chr&gt; &lt;chr&gt;         &lt;chr&gt;    &lt;chr&gt;                                
##  1  1001 y     1/0/T/F/TRUE… 2015-01… '/Library/Frameworks/R.framework/Ver…
##  2  1002 y     1/0/T/F/TRUE… 2018-05… '/Library/Frameworks/R.framework/Ver…
##  3  1003 y     1/0/T/F/TRUE… 2015-09… '/Library/Frameworks/R.framework/Ver…
##  4  1004 y     1/0/T/F/TRUE… 2012-11… '/Library/Frameworks/R.framework/Ver…
##  5  1005 y     1/0/T/F/TRUE… 2020-01… '/Library/Frameworks/R.framework/Ver…
##  6  1006 y     1/0/T/F/TRUE… 2016-04… '/Library/Frameworks/R.framework/Ver…
##  7  1007 y     1/0/T/F/TRUE… 2011-05… '/Library/Frameworks/R.framework/Ver…
##  8  1008 y     1/0/T/F/TRUE… 2020-07… '/Library/Frameworks/R.framework/Ver…
##  9  1009 y     1/0/T/F/TRUE… 2011-04… '/Library/Frameworks/R.framework/Ver…
## 10  1010 y     1/0/T/F/TRUE… 2010-05… '/Library/Frameworks/R.framework/Ver…
## # … with 990 more rows
```
OK, let's see which columns cause trouble:

```r
p %$% table(col)
```

```
## col
##    y 
## 1000
```
Looks like the problem occurs only in the `x` column.

---
name: readr_problems_fixing
## Fixing Problematic Columns
So, how can we fix the problematic columns?

1. We can explicitely tell what parser to use:

```r
tricky_dataset &lt;- read_csv(readr_example('challenge.csv'),
                           col_types = cols(x = col_double(),
                                            y = col_character()
                                            )
                                                      )
tricky_dataset %&gt;% tail(n = 5)
```

```
## # A tibble: 5 x 2
##       x y         
##   &lt;dbl&gt; &lt;chr&gt;     
## 1 0.164 2018-03-29
## 2 0.472 2014-08-04
## 3 0.718 2015-08-16
## 4 0.270 2020-02-04
## 5 0.608 2019-01-06
```
As you can see, we can still do better by parsing the `y` column as *date*, not as *character*. 

---
name: readr_problems_fixing2
## Fixing Problematic Columns cted.
But knowing that the parser is guessed based on the first 1000 lines, we can see what sits past the 1000-th line in the data:

```r
tricky_dataset %&gt;% head(n = 1002) %&gt;% tail(n = 4)
```

```
## # A tibble: 4 x 2
##          x y         
##      &lt;dbl&gt; &lt;chr&gt;     
## 1 4569     &lt;NA&gt;      
## 2 4548     &lt;NA&gt;      
## 3    0.238 2015-01-16
## 4    0.412 2018-05-18
```
It seems, we were very unlucky, because up till 1000-th line there are only integers in the x column and `NA`s in the y column so the parser cannot be guessed correctly. To fix this:

```r
tricky_dataset &lt;- read_csv(readr_example('challenge.csv'),
                           guess_max = 1001)
```

```
## Parsed with column specification:
## cols(
##   x = col_double(),
##   y = col_date(format = "")
## )
```

---
name: readr_writing
## Writing to a File
The `readr` package also provides functions useful for writing tibbled data into a file:

* `write_csv()`
* `write_tsv()`
* `write_excel_csv()`

They **always** save:

* text in UTF-8,
* dates in ISO8601

But saving in csv (or tsv) does mean you loose information about the type of data in particular columns. You can avoid this by using:

* `write_rds()` and `read_rds()` to read/write objects in R binary rds format,
* use `write_feather()` and `read_feather()` from package `feather` to read/write objects in a fast binary format that other programming languages can access.

---
name: basic_data_transformations
## Basic Data Transformations with `dplyr`

Let us create a tibble:

```r
bijou &lt;- as_tibble(diamonds) %&gt;% head(n = 100)
bijou
```

```
## # A tibble: 100 x 10
##    carat cut       color clarity depth table price     x     y     z
##    &lt;dbl&gt; &lt;ord&gt;     &lt;ord&gt; &lt;ord&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
##  1 0.23  Ideal     E     SI2      61.5    55   326  3.95  3.98  2.43
##  2 0.21  Premium   E     SI1      59.8    61   326  3.89  3.84  2.31
##  3 0.23  Good      E     VS1      56.9    65   327  4.05  4.07  2.31
##  4 0.290 Premium   I     VS2      62.4    58   334  4.2   4.23  2.63
##  5 0.31  Good      J     SI2      63.3    58   335  4.34  4.35  2.75
##  6 0.24  Very Good J     VVS2     62.8    57   336  3.94  3.96  2.48
##  7 0.24  Very Good I     VVS1     62.3    57   336  3.95  3.98  2.47
##  8 0.26  Very Good H     SI1      61.9    55   337  4.07  4.11  2.53
##  9 0.22  Fair      E     VS2      65.1    61   337  3.87  3.78  2.49
## 10 0.23  Very Good H     VS1      59.4    61   338  4     4.05  2.39
## # … with 90 more rows
```

.center[
  &lt;img src="assets_tidyverse/diamonds.png", style="height:200px"&gt;
]
---
name: filter
## Picking Observations using `filter()`

```r
bijou %&gt;% filter(cut == 'Ideal' | cut == 'Premium', carat &gt;= 0.23) %&gt;%
  head(n = 5)
```

```
## # A tibble: 5 x 10
##   carat cut     color clarity depth table price     x     y     z
##   &lt;dbl&gt; &lt;ord&gt;   &lt;ord&gt; &lt;ord&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1 0.23  Ideal   E     SI2      61.5    55   326  3.95  3.98  2.43
## 2 0.290 Premium I     VS2      62.4    58   334  4.2   4.23  2.63
## 3 0.23  Ideal   J     VS1      62.8    56   340  3.93  3.9   2.46
## 4 0.31  Ideal   J     SI2      62.2    54   344  4.35  4.37  2.71
## 5 0.32  Premium E     I1       60.9    58   345  4.38  4.42  2.68
```
Be careful with floating point comparisons! Also, rows with comparison resulting in `NA` are skipped by default!

```r
bijou %&gt;% filter(near(0.23, carat) | is.na(carat)) %&gt;%
  head(n = 5)
```

```
## # A tibble: 5 x 10
##   carat cut       color clarity depth table price     x     y     z
##   &lt;dbl&gt; &lt;ord&gt;     &lt;ord&gt; &lt;ord&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1  0.23 Ideal     E     SI2      61.5    55   326  3.95  3.98  2.43
## 2  0.23 Good      E     VS1      56.9    65   327  4.05  4.07  2.31
## 3  0.23 Very Good H     VS1      59.4    61   338  4     4.05  2.39
## 4  0.23 Ideal     J     VS1      62.8    56   340  3.93  3.9   2.46
## 5  0.23 Very Good E     VS2      63.8    55   352  3.85  3.92  2.48
```

---
name: arrange
## Rearranging Observations using `arrange()`

```r
bijou %&gt;% arrange(cut, carat, desc(price))
```

```
## # A tibble: 100 x 10
##    carat cut   color clarity depth table price     x     y     z
##    &lt;dbl&gt; &lt;ord&gt; &lt;ord&gt; &lt;ord&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
##  1  0.22 Fair  E     VS2      65.1    61   337  3.87  3.78  2.49
##  2  0.86 Fair  E     SI2      55.1    69  2757  6.45  6.33  3.52
##  3  0.96 Fair  F     SI2      66.3    62  2759  6.27  5.95  4.07
##  4  0.23 Good  F     VS1      58.2    59   402  4.06  4.08  2.37
##  5  0.23 Good  E     VS1      64.1    59   402  3.83  3.85  2.46
##  6  0.23 Good  E     VS1      56.9    65   327  4.05  4.07  2.31
##  7  0.26 Good  E     VVS1     57.9    60   554  4.22  4.25  2.45
##  8  0.26 Good  D     VS2      65.2    56   403  3.99  4.02  2.61
##  9  0.26 Good  D     VS1      58.4    63   403  4.19  4.24  2.46
## 10  0.3  Good  H     SI1      63.7    57   554  4.28  4.26  2.72
## # … with 90 more rows
```
The `NA`s always end up at the end of the rearranged tibble.

---
name: select
## Selecting Variables with `select()`
Simple `select` with a range:

```r
bijou %&gt;% select(color, clarity, x:z) %&gt;% head(n = 5)
```

```
## # A tibble: 5 x 5
##   color clarity     x     y     z
##   &lt;ord&gt; &lt;ord&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1 E     SI2      3.95  3.98  2.43
## 2 E     SI1      3.89  3.84  2.31
## 3 E     VS1      4.05  4.07  2.31
## 4 I     VS2      4.2   4.23  2.63
## 5 J     SI2      4.34  4.35  2.75
```
--
Exclusive `select`:

```r
bijou %&gt;% select(-(x:z)) %&gt;% head(n = 5)
```

```
## # A tibble: 5 x 7
##   carat cut     color clarity depth table price
##   &lt;dbl&gt; &lt;ord&gt;   &lt;ord&gt; &lt;ord&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;int&gt;
## 1 0.23  Ideal   E     SI2      61.5    55   326
## 2 0.21  Premium E     SI1      59.8    61   326
## 3 0.23  Good    E     VS1      56.9    65   327
## 4 0.290 Premium I     VS2      62.4    58   334
## 5 0.31  Good    J     SI2      63.3    58   335
```

---
name: select2
## Selecting Variables with `select()` cted.
`rename` is a variant of `select`, here used with `everything()` to move `x` to the beginning and rename it to `var_x` 

```r
bijou %&gt;% rename(var_x = x) %&gt;% head(n = 5)
```

```
## # A tibble: 5 x 10
##   carat cut     color clarity depth table price var_x     y     z
##   &lt;dbl&gt; &lt;ord&gt;   &lt;ord&gt; &lt;ord&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1 0.23  Ideal   E     SI2      61.5    55   326  3.95  3.98  2.43
## 2 0.21  Premium E     SI1      59.8    61   326  3.89  3.84  2.31
## 3 0.23  Good    E     VS1      56.9    65   327  4.05  4.07  2.31
## 4 0.290 Premium I     VS2      62.4    58   334  4.2   4.23  2.63
## 5 0.31  Good    J     SI2      63.3    58   335  4.34  4.35  2.75
```
--
use `everything()` to bring some columns to the front:

```r
bijou %&gt;% select(x:z, everything()) %&gt;% head(n = 5)
```

```
## # A tibble: 5 x 10
##       x     y     z carat cut     color clarity depth table price
##   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;ord&gt;   &lt;ord&gt; &lt;ord&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;int&gt;
## 1  3.95  3.98  2.43 0.23  Ideal   E     SI2      61.5    55   326
## 2  3.89  3.84  2.31 0.21  Premium E     SI1      59.8    61   326
## 3  4.05  4.07  2.31 0.23  Good    E     VS1      56.9    65   327
## 4  4.2   4.23  2.63 0.290 Premium I     VS2      62.4    58   334
## 5  4.34  4.35  2.75 0.31  Good    J     SI2      63.3    58   335
```

---
name: mutate
## Create/alter new Variables with `mutate` 

```r
bijou %&gt;% mutate(p = x + z, q = p + y) %&gt;% select(-(depth:price)) %&gt;% head(n = 5)
```

```
## # A tibble: 5 x 9
##   carat cut     color clarity     x     y     z     p     q
##   &lt;dbl&gt; &lt;ord&gt;   &lt;ord&gt; &lt;ord&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1 0.23  Ideal   E     SI2      3.95  3.98  2.43  6.38  10.4
## 2 0.21  Premium E     SI1      3.89  3.84  2.31  6.2   10.0
## 3 0.23  Good    E     VS1      4.05  4.07  2.31  6.36  10.4
## 4 0.290 Premium I     VS2      4.2   4.23  2.63  6.83  11.1
## 5 0.31  Good    J     SI2      4.34  4.35  2.75  7.09  11.4
```
--
or with `transmute` (only the transformed variables will be retained)

```r
bijou %&gt;% transmute(carat, cut, sum = x + y + z) %&gt;% head(n = 5)
```

```
## # A tibble: 5 x 3
##   carat cut       sum
##   &lt;dbl&gt; &lt;ord&gt;   &lt;dbl&gt;
## 1 0.23  Ideal    10.4
## 2 0.21  Premium  10.0
## 3 0.23  Good     10.4
## 4 0.290 Premium  11.1
## 5 0.31  Good     11.4
```

---
name: grouped_summaries
## Group and Summarize

```r
bijou %&gt;% group_by(cut) %&gt;% summarize(max_price = max(price),
                                      mean_price = mean(price),
                                      min_price = min(price))
```

```
## # A tibble: 5 x 4
##   cut       max_price mean_price min_price
##   &lt;ord&gt;         &lt;int&gt;      &lt;dbl&gt;     &lt;int&gt;
## 1 Fair           2759      1951        337
## 2 Good           2759       661.       327
## 3 Very Good      2760       610.       336
## 4 Premium        2760       569.       326
## 5 Ideal          2757       693.       326
```
--

```r
bijou %&gt;% 
  group_by(cut, color) %&gt;% 
  summarize(max_price = max(price), 
            mean_price = mean(price), 
            min_price = min(price)) %&gt;% head(n = 5)
```

```
## # A tibble: 5 x 5
## # Groups:   cut [2]
##   cut   color max_price mean_price min_price
##   &lt;ord&gt; &lt;ord&gt;     &lt;int&gt;      &lt;dbl&gt;     &lt;int&gt;
## 1 Fair  E          2757      1547        337
## 2 Fair  F          2759      2759       2759
## 3 Good  D           403       403        403
## 4 Good  E          2759      1010.       327
## 5 Good  F          2759      1580.       402
```

---
name: other_data_manipulations
## Other data manipulation tips

```r
bijou %&gt;% group_by(cut) %&gt;% summarize(count = n())
```

```
## # A tibble: 5 x 2
##   cut       count
##   &lt;ord&gt;     &lt;int&gt;
## 1 Fair          3
## 2 Good         18
## 3 Very Good    38
## 4 Premium      22
## 5 Ideal        19
```
--
When you need to regroup within the same pipe, use `ungroup()`.
---
name: concept_of_tidy_data

## The Concept of Tidy Data
Data are tidy *sensu Wickham* if:
* each and every observation is represented as exactly one row,
* each and every variable is represented by exactly one column,
* thus each data table cell contains only one value.
&lt;img src="assets_tidyverse/tidy_data.png" width="2560" style="display: block; margin: auto auto auto 0;" /&gt;

Usually data are untidy in only one way. However, if you are unlucky, they are really untidy and thus a pain to work with...
---
name: tidy_data

## Tidy Data
&lt;img src="assets_tidyverse/tidy_data.png" width="2560" style="display: block; margin: auto auto auto 0;" /&gt;
--
.center[**Are these data tidy?**]

.pull-left-70[
&lt;table class="table table-striped table-hover table-responsive table-condensed" style="width: auto !important; "&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:center;"&gt; Sepal.Length &lt;/th&gt;
   &lt;th style="text-align:center;"&gt; Sepal.Width &lt;/th&gt;
   &lt;th style="text-align:center;"&gt; Petal.Length &lt;/th&gt;
   &lt;th style="text-align:center;"&gt; Petal.Width &lt;/th&gt;
   &lt;th style="text-align:center;"&gt; Species &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:center;"&gt; 5.1 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 3.5 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 1.4 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 0.2 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; setosa &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:center;"&gt; 4.9 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 3.0 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 1.4 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 0.2 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; setosa &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:center;"&gt; 4.7 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 3.2 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 1.3 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 0.2 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; setosa &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;
]
--
.pull-right-30[
&lt;table class="table table-striped table-hover table-responsive table-condensed" style="width: auto !important; "&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:center;"&gt; Species &lt;/th&gt;
   &lt;th style="text-align:center;"&gt; variable &lt;/th&gt;
   &lt;th style="text-align:center;"&gt; value &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:center;"&gt; setosa &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; Sepal.Length &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 5.1 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:center;"&gt; setosa &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; Sepal.Length &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 4.9 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:center;"&gt; setosa &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; Sepal.Length &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 4.7 &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;
]
&lt;br&gt;&amp;nbsp;&lt;hr&gt;&lt;br&gt;

--
.pull-left-50[
&lt;table class="table table-striped table-hover table-responsive table-condensed" style="width: auto !important; "&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:center;"&gt; Sepal.L.W &lt;/th&gt;
   &lt;th style="text-align:center;"&gt; Petal.L.W &lt;/th&gt;
   &lt;th style="text-align:center;"&gt; Species &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:center;"&gt; 5.1/3.5 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 1.4/0.2 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; setosa &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:center;"&gt; 4.9/3 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 1.4/0.2 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; setosa &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:center;"&gt; 4.7/3.2 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 1.3/0.2 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; setosa &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;
]
--
.pull-right-50[
&lt;table class="table table-striped table-hover table-responsive table-condensed" style="width: auto !important; "&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Sepal.Length &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 5.1 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 4.9 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 4.7 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 4.6 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Sepal.Width &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 3.5 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 3.0 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 3.2 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 3.1 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Petal.Length &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 1.4 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 1.4 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 1.3 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 1.5 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Petal.Width &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 0.2 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 0.2 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 0.2 &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; 0.2 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Species &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; setosa &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; setosa &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; setosa &lt;/td&gt;
   &lt;td style="text-align:center;"&gt; setosa &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;
]

---
name: tidying_data_gather
## Tidying Data with `gather`

If some of your column names are actually values of a variable, use `gather`:



```r
bijou2 %&gt;% head(n = 5)
```

```
## # A tibble: 5 x 3
##   cut     `2008` `2009`
##   &lt;ord&gt;    &lt;int&gt;  &lt;dbl&gt;
## 1 Ideal      326   333.
## 2 Premium    326   333.
## 3 Good       327   334.
## 4 Premium    334   341.
## 5 Good       335   342.
```

```r
bijou2 %&gt;% 
  gather(`2008`, `2009`, key = 'year', value = 'price') %&gt;% 
  head(n = 5)
```

```
## # A tibble: 5 x 3
##   cut     year  price
##   &lt;ord&gt;   &lt;chr&gt; &lt;dbl&gt;
## 1 Ideal   2008    326
## 2 Premium 2008    326
## 3 Good    2008    327
## 4 Premium 2008    334
## 5 Good    2008    335
```

---
name: tidying_data_spread
## Tidying Data with `spread`

If some of your observations are scattered across many rows, use `gather`:



```r
bijou3
```

```
## # A tibble: 9 x 5
##   cut     price clarity dimension measurement
##   &lt;ord&gt;   &lt;int&gt; &lt;ord&gt;   &lt;chr&gt;           &lt;dbl&gt;
## 1 Ideal     326 SI2     x                3.95
## 2 Premium   326 SI1     x                3.89
## 3 Good      327 VS1     x                4.05
## 4 Ideal     326 SI2     y                3.98
## 5 Premium   326 SI1     y                3.84
## 6 Good      327 VS1     y                4.07
## 7 Ideal     326 SI2     z                2.43
## 8 Premium   326 SI1     z                2.31
## 9 Good      327 VS1     z                2.31
```

```r
bijou3 %&gt;% 
  spread(key=dimension, value=measurement) %&gt;% 
  head(n = 5)
```

```
## # A tibble: 3 x 6
##   cut     price clarity     x     y     z
##   &lt;ord&gt;   &lt;int&gt; &lt;ord&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1 Good      327 VS1      4.05  4.07  2.31
## 2 Premium   326 SI1      3.89  3.84  2.31
## 3 Ideal     326 SI2      3.95  3.98  2.43
```

---
name: tidying_data_separate
## Tidying Data with `separate`

If some of your columns contain more than one value, use `separate`:



```r
bijou4
```

```
## # A tibble: 5 x 4
##   cut     price clarity dim           
##   &lt;ord&gt;   &lt;int&gt; &lt;ord&gt;   &lt;chr&gt;         
## 1 Ideal     326 SI2     3.95/3.98/2.43
## 2 Premium   326 SI1     3.89/3.84/2.31
## 3 Good      327 VS1     4.05/4.07/2.31
## 4 Premium   334 VS2     4.2/4.23/2.63 
## 5 Good      335 SI2     4.34/4.35/2.75
```

```r
bijou4 %&gt;% 
  separate(dim, into = c("x", "y", "z"), sep = "/", convert = T)
```

```
## # A tibble: 5 x 6
##   cut     price clarity     x     y     z
##   &lt;ord&gt;   &lt;int&gt; &lt;ord&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1 Ideal     326 SI2      3.95  3.98  2.43
## 2 Premium   326 SI1      3.89  3.84  2.31
## 3 Good      327 VS1      4.05  4.07  2.31
## 4 Premium   334 VS2      4.2   4.23  2.63
## 5 Good      335 SI2      4.34  4.35  2.75
```

---
name: tidying_data_separate
## Tidying Data with `unite`

If some of your columns contain more than one value, use `separate`:



```r
bijou5
```

```
## # A tibble: 5 x 7
##   cut     price clarity_prefix clarity_suffix     x     y     z
##   &lt;ord&gt;   &lt;int&gt; &lt;chr&gt;          &lt;chr&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1 Ideal     326 SI             2               3.95  3.98  2.43
## 2 Premium   326 SI             1               3.89  3.84  2.31
## 3 Good      327 VS             1               4.05  4.07  2.31
## 4 Premium   334 VS             2               4.2   4.23  2.63
## 5 Good      335 SI             2               4.34  4.35  2.75
```

```r
bijou5 %&gt;% unite(clarity, clarity_prefix, clarity_suffix, sep='')
```

```
## # A tibble: 5 x 6
##   cut     price clarity     x     y     z
##   &lt;ord&gt;   &lt;int&gt; &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1 Ideal     326 SI2      3.95  3.98  2.43
## 2 Premium   326 SI1      3.89  3.84  2.31
## 3 Good      327 VS1      4.05  4.07  2.31
## 4 Premium   334 VS2      4.2   4.23  2.63
## 5 Good      335 SI2      4.34  4.35  2.75
```
**Note:** that `sep` is here interpreted as the position to split on. It can also be a *regular expression* or a delimiting string/character. Pretty flexible approach!

---
name: missing_complete
## Completing Missing Values Using `complete`


```r
bijou %&gt;% head(n = 10) %&gt;% 
  select(cut, clarity, price) %&gt;% 
  mutate(continent = sample(c('AusOce', 'Eur'), 
                            size = 10, 
                            replace = T)) -&gt; missing_stones
```

```r
missing_stones %&gt;% complete(cut, continent)
```

```
## # A tibble: 12 x 4
##    cut       continent clarity price
##    &lt;ord&gt;     &lt;chr&gt;     &lt;ord&gt;   &lt;int&gt;
##  1 Fair      AusOce    VS2       337
##  2 Fair      Eur       &lt;NA&gt;       NA
##  3 Good      AusOce    SI2       335
##  4 Good      Eur       VS1       327
##  5 Very Good AusOce    VVS2      336
##  6 Very Good AusOce    SI1       337
##  7 Very Good AusOce    VS1       338
##  8 Very Good Eur       VVS1      336
##  9 Premium   AusOce    SI1       326
## 10 Premium   Eur       VS2       334
## 11 Ideal     AusOce    &lt;NA&gt;       NA
## 12 Ideal     Eur       SI2       326
```

---
name: joins
## Combining Datasets
Often, we need to combine a number of data tables (relational data) to get the full picture of the data. Here different types of *joins* come to help:
* *mutating joins* that add new variables to data table `A` based on matching observations (rows) from data table `B`,
* *filtering joins* that filter observations from data table `A` based on whether they match observations in data table `B`,
* *set operations* that treat observations in `A` and `B` as elements of a set.

--

Let us create two example tibbles that share a key:
.pull-left-50[

```r
A &lt;- tribble(
  ~key, ~x,
  'a', 'A1',
  'b', 'A2',
  'c', 'A3',
  'e','A4'
)
```
&lt;table&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:left;"&gt; key &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; x &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; a &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; A1 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; b &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; A2 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; c &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; A3 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; e &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; A4 &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;
]
.pull-right-50[

```r
B &lt;- tribble(
  ~key, ~y,
  'a', 'B1',
  'b', NA,
  'c', 'B3',
  'd','B4'
)
```
&lt;table&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:left;"&gt; key &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; y &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; a &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; B1 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; b &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; NA &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; c &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; B3 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; d &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; B4 &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;
]

---
name: inner_join
## The Joins Family

.pull-left-50[
The `inner join`:

```r
A %&gt;% inner_join(B, by = 'key')
# All non-matching rows are dropped!
```

```
## # A tibble: 3 x 3
##   key   x     y    
##   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;
## 1 a     A1    B1   
## 2 b     A2    &lt;NA&gt; 
## 3 c     A3    B3
```
]
--
.pull-right-50[
The `left_join`:

```r
A %&gt;% left_join(B, by = 'key')
```

```
## # A tibble: 4 x 3
##   key   x     y    
##   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;
## 1 a     A1    B1   
## 2 b     A2    &lt;NA&gt; 
## 3 c     A3    B3   
## 4 e     A4    &lt;NA&gt;
```
]
--
&lt;br&gt;
.pull-left-50[
The `right_join`:

```r
A %&gt;% right_join(B, by = 'key')
```

```
## # A tibble: 4 x 3
##   key   x     y    
##   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;
## 1 a     A1    B1   
## 2 b     A2    &lt;NA&gt; 
## 3 c     A3    B3   
## 4 d     &lt;NA&gt;  B4
```
]
--
.pull-right-50[
The `full_join`:

```r
A %&gt;% full_join(B, by = 'key')
```

```
## # A tibble: 5 x 3
##   key   x     y    
##   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;
## 1 a     A1    B1   
## 2 b     A2    &lt;NA&gt; 
## 3 c     A3    B3   
## 4 e     A4    &lt;NA&gt; 
## 5 d     &lt;NA&gt;  B4
```
]

---
name: more_tidyverse
## Some Other Friends
* `stringr` for string manipulation and regular expressions,
* `forcats` for working with factors,
* `lubridate` for working with dates.
---
name: end-slide
class: end-slide

# Thank you

---
name: purrr_map
## Using `map` Functions from `purrr`

Base R `apply` functions have their `tidyverse` counterparts.

```r
cars &lt;- as_tibble(mtcars)
cars %&gt;% head(n = 5)
```

```
## # A tibble: 5 x 11
##     mpg   cyl  disp    hp  drat    wt  qsec    vs    am  gear  carb
##   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1  21       6   160   110  3.9   2.62  16.5     0     1     4     4
## 2  21       6   160   110  3.9   2.88  17.0     0     1     4     4
## 3  22.8     4   108    93  3.85  2.32  18.6     1     1     4     1
## 4  21.4     6   258   110  3.08  3.22  19.4     1     0     3     1
## 5  18.7     8   360   175  3.15  3.44  17.0     0     0     3     2
```
The `map` function:

```r
cars %&gt;% select(disp, hp) %&gt;% map(mean)
```

```
## $disp
## [1] 230.7219
## 
## $hp
## [1] 146.6875
```

---
name: purrr_map_family
## Different Members of the `map` Family
* `map()` -- returns a list,
* `map_lgl()` -- returns a logical vector,
* `map_int()` -- returns a vector of integers,
* `map_dbl()` -- returns a vector of doubles,
* `map_chr()` -- returns a vector of characters.

---
name: purrr_shortcut_anonymous
## Anonymous Functions in `purrr`
.pull-left-50[
base-R

```r
models &lt;- cars %&gt;% 
  split(.$cyl) %&gt;%
  map(function(dat) 
    lm(mpg ~ wt, data = dat))
```
Now, make summary for each model:

```r
models %&gt;% 
  map(summary) %&gt;%
  map_dbl(~.$r.squared)
```

```
##         4         6         8 
## 0.5086326 0.4645102 0.4229655
```
]
.pull-right-50[
`purrr`

```r
models &lt;- cars %&gt;% 
  split(.$cyl) %&gt;%
  map(~lm(mpg ~ wt, data = .))
```
Now, make summary for each model using even simpler syntax:

```r
models %&gt;% 
  map(summary) %&gt;%
  map_dbl("r.squared")
```

```
##         4         6         8 
## 0.5086326 0.4645102 0.4229655
```
]

---
name: purrr_safely
## Possibly Quiet and Safe
How to deal with errors in `purrr`:

* `safely()` -- result is a list with 2 elements:

.pull-left-50[
  + `result` contains NULL if error occured, the result otherwise,
  
  ```r
  safe_sqrt &lt;- safely(sqrt)
  safe_sqrt(4) %&gt;% str()
  ```
  
  ```
  ## List of 2
  ##  $ result: num 2
  ##  $ error : NULL
  ```
]
.pull-right-50[
  + `error` contains NULL if no error occured, error object otherwise
  
  ```r
  safe_sqrt &lt;- safely(sqrt)
  safe_sqrt('zebra') %&gt;% str()
  ```
  
  ```
  ## List of 2
  ##  $ result: NULL
  ##  $ error :List of 2
  ##   ..$ message: chr "non-numeric argument to mathematical function"
  ##   ..$ call   : language .Primitive("sqrt")(x)
  ##   ..- attr(*, "class")= chr [1:3] "simpleError" "error" "condition"
  ```
]

---
name: purrr_possibly
## Possibly Quiet and Safe cted.

* `possibly()` let's you define what value to return upon error:

```r
tst &lt;- list(4, 7, 'test')
tst %&gt;% map_dbl(possibly(sqrt, NA_real_))
```

```
## [1] 2.000000 2.645751       NA
```
--
* `quietly()` -- it captures output, messages and warnings and returns it as a list:

```r
x &lt;- list(-1, 1)
x %&gt;% map(quietly(log)) %&gt;% str()
```

```
## List of 2
##  $ :List of 4
##   ..$ result  : num NaN
##   ..$ output  : chr ""
##   ..$ warnings: chr "NaNs produced"
##   ..$ messages: chr(0) 
##  $ :List of 4
##   ..$ result  : num 0
##   ..$ output  : chr ""
##   ..$ warnings: chr(0) 
##   ..$ messages: chr(0)
```

---
name: more_on_map
## More on `map` Functions
What if one wants to map over more than one argument? 

```r
means &lt;- c(22, 32, 42)
std_devs &lt;- c(2.5, 5, 10)
my_rnorms &lt;- map2(means, std_devs, rnorm, n = 100)
```
--
&lt;img src="tidyverse_presentation_files/figure-html/unnamed-chunk-55-1.png" style="display: block; margin: auto;" /&gt;
--

```r
 my_rnorms %&gt;% 
  setNames(LETTERS[1:length(means)]) %&gt;% 
  as_tibble() %&gt;% 
  gather(LETTERS[1]:LETTERS[length(means)], key='run', value='num') %&gt;% 
  ggplot(mapping = aes(x = num)) + 
  geom_density() + 
  facet_grid(~ run) + 
  theme_bw()
```

---
name: more_on_map
## Mapping with Even More Arguments

```r
param_sets &lt;- tribble(
  ~mean, ~sd, ~n,
  22, 2.5, 50,
  32, 5, 100,
  42, 10, 250
)
```
--
&lt;table&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:right;"&gt; mean &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; sd &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; n &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:right;"&gt; 22 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 2.5 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 50 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:right;"&gt; 32 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 5.0 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 100 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:right;"&gt; 42 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 10.0 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 250 &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;


--

```r
param_sets %&gt;% pmap(rnorm) %&gt;% str()
```

```
## List of 3
##  $ : num [1:50] 24.7 23.4 23.3 22 25.4 ...
##  $ : num [1:100] 23.6 32.8 24.7 38.5 37.8 ...
##  $ : num [1:250] 27.6 43.2 57.1 33.9 34 ...
```
---
name: invoke_map 
## Invoking Different Functions

```r
param_sets &lt;- tribble(
  ~f,       ~params,
  "runif", list(min = -1, max = 1),
  "rnorm", list(mean = 32, sd = 2),
  "rpois", list(lambda = 10)
)
result &lt;- param_sets %&gt;% 
  mutate(call_result = invoke_map(f, params, n = 100))
```
--
&lt;img src="tidyverse_presentation_files/figure-html/unnamed-chunk-60-1.png" style="display: block; margin: auto;" /&gt;

---
name: walk
## Let's Take a `walk` to the Printing House
What if you want to map a function for its side-effects?

```r
list(runif(10), rnorm(10)) %&gt;% 
  walk(print) %&gt;% 
  map(`*`,5)
```

```
##  [1] 0.90274778 0.28083973 0.13133800 0.45774198 0.08027158 0.36694898
##  [7] 0.77141715 0.82984949 0.23660029 0.39106727
##  [1]  1.04379058  0.57891213 -1.25773084 -1.86956841  0.66966623
##  [6]  0.04672486 -0.06080231  0.99140409 -0.94845375 -1.73384066
## [[1]]
##  [1] 4.5137389 1.4041986 0.6566900 2.2887099 0.4013579 1.8347449 3.8570857
##  [8] 4.1492474 1.1830014 1.9553364
## 
## [[2]]
##  [1]  5.2189529  2.8945607 -6.2886542 -9.3478420  3.3483311  0.2336243
##  [7] -0.3040115  4.9570204 -4.7422687 -8.6692033
```
* `walk2()`
* `pwalk()`

---
name: some_every
## Predicate Functions
* keep all elements fulfilling a condition:

```r
iris %&gt;% keep(is.factor) %&gt;% str()
```

```
## 'data.frame':	150 obs. of  1 variable:
##  $ Species: Factor w/ 3 levels "setosa","versicolor",..: 1 1 1 1 1 1 1 1 1 1 ...
```
* discard all elements fulfilling a condition:

```r
iris$Petal.Length %&gt;% discard(~ . &gt;= 2) %&gt;% str()
```

```
##  num [1:50] 1.4 1.4 1.3 1.5 1.4 1.7 1.4 1.5 1.4 1.5 ...
```
* `some()` and `every()`
* `detect()` and `detect_index()`

```r
10:1 %&gt;% detect(~ . &gt; 5)
10:1 %&gt;% detect_index(~ . &gt; 5)
```

```
## [1] 10
## [1] 1
```
* `head_while()`, `tail_while()`

---
name: more_tidyverse
## Some Other Friends
* `stringr` for string manipulation and regular expressions,
* `forcats` for working with factors,
* `lubridate` for working with dates.
---
name: end-slide
class: end-slide

# Thank you

---
name: report

## Session  

* This presentation was created in RStudio using [`remarkjs`](https://github.com/gnab/remark) framework through R package [`xaringan`](https://github.com/yihui/xaringan).
* For R Markdown, see &lt;http://rmarkdown.rstudio.com&gt;
* For R Markdown presentations, see &lt;https://rmarkdown.rstudio.com/lesson-11.html&gt;


```r
R.version
```

```
##                _                           
## platform       x86_64-apple-darwin15.6.0   
## arch           x86_64                      
## os             darwin15.6.0                
## system         x86_64, darwin15.6.0        
## status                                     
## major          3                           
## minor          5.0                         
## year           2018                        
## month          04                          
## day            23                          
## svn rev        74626                       
## language       R                           
## version.string R version 3.5.0 (2018-04-23)
## nickname       Joy in Playing
```
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="assets/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "4:3",
"highlightLanguage": "r",
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"slideNumberFormat": "RaukR 2019 • %current%/%total%"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();</script>

<script>
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
