<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Marcin Kierczak" />


<title>Object-Oriented Programming in R</title>

<script src="oop_lab_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="oop_lab_files/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="oop_lab_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="oop_lab_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="oop_lab_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="oop_lab_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="oop_lab_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="oop_lab_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="oop_lab_files/navigation-1.1/tabsets.js"></script>
<script src="oop_lab_files/navigation-1.1/codefolding.js"></script>
<link href="oop_lab_files/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="oop_lab_files/highlightjs-9.12.0/highlight.js"></script>
<link href="oop_lab_files/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="oop_lab_files/pagedtable-1.1/js/pagedtable.js"></script>
<link href="oop_lab_files/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="oop_lab_files/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="oop_lab_files/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Object-Oriented Programming in R</h1>
<h3 class="subtitle">RaukR 2019 • Advanced R for Bioinformatics</h3>
<h4 class="author">Marcin Kierczak</h4>

</div>


<p><link href="https://fonts.googleapis.com/css?family=Roboto|Source+Sans+Pro:300,400,600|Ubuntu+Mono&amp;subset=latin-ext" rel="stylesheet"></p>
<p><img src="assets/logo.svg" alt="logo_raukr" class="trlogo"></p>
<div class="abstract">
<p>During this lab, we will go through the most important features of 3 out of the 4 existing OOP systems in R:</p>
<ul>
<li>S3</li>
<li>S4</li>
<li>R6</li>
</ul>
<p>While S3 and S4 are base-R OOP systems, R6 is provided by the <code>R6</code> package. Both S3 and S4 follow the <em>functional OOP style</em> while <code>R6</code> is more similar to OOP models known from Java or C++. We will not cover the Reference Classes (a.k.a. RC) which is yet another base-R OOP model. Why? Well, it is really tricky to use and its documentation is not always complete…</p>
</div>
<p><br></p>
<div id="s3-classes" class="section level1">
<h1><span class="header-section-number">1</span> S3 Classes</h1>
<ul>
<li>What is the class of the object returned by the <code>lm()</code> function?</li>
</ul>
<pre class="r"><code>obj &lt;- lm(speed ~ dist, data = cars) 
class(obj)</code></pre>
<ul>
<li>What basic data type is it build upon?</li>
</ul>
<pre class="r"><code>typeof(obj)</code></pre>
<ul>
<li>What attributes does the object of the <code>lm</code> class have?</li>
</ul>
<pre class="r"><code>attributes(obj)</code></pre>
<ul>
<li>What is the structure of the <code>lm</code> object?</li>
</ul>
<pre class="r"><code>str(obj)</code></pre>
<ul>
<li>Does the <code>lm</code> class implement own <code>str()</code>?</li>
</ul>
<pre class="r"><code>str(obj)
# compare to:
str(unclass(obj)) 

# Answer: no</code></pre>
<ul>
<li>What is the class of a <code>tibble</code>? You have to load the <code>tidyverse</code> package and use one of the built-in datasets. e.g. <code>mtcars</code> as argument to <code>as_tibble()</code> to check the class.</li>
</ul>
<pre class="r"><code>library(tidyverse)
obj &lt;- as_tibble(mtcars)
class(obj)</code></pre>
<ul>
<li>What is the underlying data type?</li>
</ul>
<pre class="r"><code>typeof(obj)</code></pre>
<ul>
<li>Is the <code>str()</code> function used by tibbles the default <code>str()</code>?</li>
</ul>
<pre class="r"><code>str(obj)
str(unclass(obj))
# Answer: no</code></pre>
<ul>
<li>Implement your own class <code>meta_frame</code> that is a <code>data.frame</code> with additional attributes:
<ul>
<li>character attribute <code>descr</code> (a description),</li>
<li><code>creation</code> attribute which will hold the creation date.</li>
</ul></li>
</ul>
<p>Check that it works, i.e. that it has the expected attributes and that it is a separate class.</p>
<p>How do you access the <code>description</code> field?</p>
<pre class="r"><code>meta_frame &lt;- function(df, descr, creation) {
  structure(
    list(
      df = df, 
      descr = descr, 
      creation = creation
    ),
    class = &quot;meta_frame&quot;
  )
}
obj &lt;- meta_frame(mtcars, &#39;The mtcars dataset&#39;, &#39;10-06-2018&#39;)
str(obj)
obj$descr</code></pre>
<ul>
<li>Design implementation of your own S3 class <code>protein</code> that stores the following information (you will implement the class in the next point, now just think and draw on paper if you wish):
<ul>
<li>protein sequence</li>
<li>protein length</li>
<li>post-translational modification (PTM) site coords</li>
<li>PTM type that can be either of: “phosphorylation”, or “methylation”</li>
</ul></li>
<li><p>Implement the <code>protein</code> class from the previous point using the <em>constructor, validator and helper</em> design pattern. Try to reduce the number of input arguments to:</p>
<ul>
<li>a sequence and</li>
<li>a list of PTMs.</li>
</ul></li>
</ul>
<pre class="r"><code># Constructor
new_protein &lt;- function(sequence, length, ptm_site, ptm_type) {
  stopifnot(is.character(sequence))
  stopifnot(is.integer(length))
  stopifnot(is.numeric(ptm_site))
  stopifnot(is.character(ptm_type))
  
  structure(
    sequence,
    length = length,
    ptm_site = ptm_site,
    ptm_type = ptm_type,
    class = &quot;protein&quot;
  )
}

# Helper
protein &lt;- function(sequence, ptm) {
  sequence &lt;- sequence
  len &lt;- nchar(sequence)
  ptm_site &lt;- ptm[[1]]
  ptm_type = ptm[[2]]
  validate_protein(new_protein(sequence, len, ptm_site, ptm_type))
}

# Validator
validate_protein &lt;- function(x) {
  sequence &lt;- unclass(x)
  len &lt;- attr(x, &#39;length&#39;)
  ptm_site &lt;- attr(x, &#39;ptm_site&#39;)
  ptm_type &lt;- attr(x, &#39;ptm_type&#39;)
  
  if (is.na(sequence)) {
    stop(&#39;Sequence is missing!&#39;, call. = FALSE)
  }
  if (length(ptm_site) != length(ptm_type)) {
    stop(&#39;PTM type and site differ in length!&#39;, call. = FALSE)
  }
  if ((sum(!ptm_type %in% c(&#39;phosphorylation&#39;, &#39;methylation&#39;))) &gt; 0) {
    stop(&#39;Invalid PTM types detected!&#39;, call. = FALSE)
  }
  
  return(x)
}</code></pre>
<ul>
<li>What would you improve our implementation of <code>protein</code>. Is it really user-friendly?</li>
<li>Write a generic <code>print</code> for the <code>protein</code> class. It should print e.g.:</li>
</ul>
<pre><code>[1] &quot;============== Protein ==============&quot;
[1] &quot;Sequence:ARNDKLLQWYTTARD&quot;
[1] &quot;Length: 15 aa.&quot;
[1] &quot;============== PTM section ==============&quot;
[1] &quot;Site: 3&quot; &quot;Site: 5&quot;
[1] &quot;Type: phosphorylation&quot; &quot;Type: methylation&quot;  </code></pre>
<pre class="r"><code>print.protein &lt;- function(x) {
  sequence &lt;- unclass(x)
  len &lt;- attr(x, &#39;length&#39;)
  ptm_site &lt;- attr(x, &#39;ptm_site&#39;)
  ptm_type &lt;- attr(x, &#39;ptm_type&#39;)
  print(&quot;============== Protein ==============&quot;)
  print(paste0(&quot;Sequence:&quot;, sequence))
  print(paste0(&quot;Length: &quot;, len, &quot; aa.&quot;))
  print(&quot;============== PTM section ==============&quot;)
  print(paste0(&quot;Site: &quot;, ptm_site))
  print(paste0(&quot;Type: &quot;, ptm_type))
}</code></pre>
<ul>
<li>Test that the <code>protein</code> class works as it should and that generic <code>print</code> works as well.</li>
</ul>
<pre class="r"><code>my_prot &lt;- protein(&quot;ARNDKLLQWYTTARD&quot;, 
                   list(
                     site = c(3, 5), 
                     type = c(&#39;phosphorylation&#39;, &#39;methylation&#39;)
                    )
                  )
class(my_prot)
typeof(my_prot)
str(my_prot)</code></pre>
</div>
<div id="s4-classes" class="section level1">
<h1><span class="header-section-number">2</span> S4 Classes</h1>
<ul>
<li>Re-write the <code>protein</code> class in S4 (with validation!).</li>
</ul>
<pre class="r"><code># Generator
.protein &lt;- setClass(&#39;protein&#39;,
                     slots = c(
                       sequence = &#39;character&#39;,
                       length = &#39;numeric&#39;,
                       ptm_site = &#39;numeric&#39;,
                       ptm_type = &#39;character&#39;
                      )
                    )

# Constructor
protein &lt;- function(sequence, ptm) {
  sequence &lt;- sequence
  len &lt;- nchar(sequence)
  
  ptm_site &lt;- ptm[[1]]
  ptm_type &lt;- ptm[[2]]
  
  if (is.na(sequence)) {
    stop(&#39;Sequence is missing!&#39;, call. = FALSE)
  }
  if (length(ptm_site) != length(ptm_type)) {
    stop(&#39;PTM type and site differ in length!&#39;, call. = FALSE)
  }
  if ((sum(!ptm_type %in% c(&#39;phosphorylation&#39;, &#39;methylation&#39;))) &gt; 0) {
    stop(&#39;Invalid PTM types detected!&#39;, call. = FALSE)
  }
  
  pt &lt;- .protein(sequence = sequence, 
           length = len, 
           ptm_site = ptm_site, 
           ptm_type = ptm_type)
  
  return(pt)
}</code></pre>
<ul>
<li>Create an S4 object of the <code>protein</code> class and check whether it works.</li>
</ul>
<pre class="r"><code>my_prot &lt;- protein(&quot;ARNDKLLQWYTTARD&quot;, 
                   list(
                     site = c(3, 5), 
                     type = c(&#39;phosphorylation&#39;, &#39;methylation&#39;)
                    )
                  )
class(my_prot)
typeof(my_prot)
str(my_prot)</code></pre>
<ul>
<li>Implement the generic <code>print</code> using S4 and check that it works.</li>
</ul>
<pre class="r"><code>setMethod(&#39;print&#39;, &#39;protein&#39;, 
  function(x) {
    sequence &lt;- x@sequence
    len &lt;- x@length
    ptm_site &lt;- x@ptm_site
    ptm_type &lt;- x@ptm_type
    print(&quot;============== Protein ==============&quot;)
    print(paste0(&quot;Sequence:&quot;, sequence))
    print(paste0(&quot;Length: &quot;, len, &quot; aa.&quot;))
    print(&quot;============== PTM section ==============&quot;)
    print(paste0(&quot;Site: &quot;, ptm_site))
    print(paste0(&quot;Type: &quot;, ptm_type))
})

print(my_prot)</code></pre>
<ul>
<li>Implement a new S4 class <code>ext_protein</code> that extends <code>protein</code> with 3 slots:
<ul>
<li>feature type</li>
<li>feature position</li>
<li>feature value</li>
</ul></li>
</ul>
<pre class="r"><code>.ext_protein &lt;- setClass(&#39;ext_protein&#39;, 
                         contains = c(&#39;protein&#39;),
                         slots = c(
                           prot = &#39;protein&#39;,
                           feature_type = &#39;character&#39;,
                           feature_position = &#39;numeric&#39;,
                           feature_value = &#39;character&#39;
                         ))
my_ext_prot &lt;- .ext_protein(prot = my_prot,
                           feature_type = &#39;modification&#39;,
                           feature_position = 11,
                           feature_value = &#39;absent&#39;
                  )
class(my_ext_prot)
typeof(my_ext_prot)
str(my_ext_prot)</code></pre>
</div>
<div id="r6-classes" class="section level1">
<h1><span class="header-section-number">3</span> R6 Classes</h1>
<ul>
<li>Install and load the <code>R6</code> package,</li>
<li>Implement the <code>protein</code> class using R6 model and check that it works as expected:</li>
</ul>
<pre class="r"><code>library(R6)
protein &lt;- R6Class(classname = &#39;protein&#39;,
                        public = list(
                          seq = NA,
                          length = NULL,
                          ptm_site = NA,
                          ptm_type = NA,
                          initialize = function(seq = NA, ptm = NA) {
                            self$seq &lt;- seq
                            self$length &lt;- nchar(self$seq)
                            self$ptm_site &lt;- ptm[[1]]
                            self$ptm_type &lt;- ptm[[2]]
                            # Check types
                            stopifnot(is.character(seq))
                            
                            #Validate
                            if (is.na(self$seq)) {
                              stop(&#39;Sequence is missing!&#39;, call. = FALSE)
                            }
                            if (length(self$ptm_site) != length(self$ptm_type)) {
                              stop(&#39;PTM type and site differ in length!&#39;, call. = FALSE)
                            }
                            if ((sum(!self$ptm_type %in% c(&#39;phosphorylation&#39;, &#39;methylation&#39;))) &gt; 0) {
    stop(&#39;Invalid PTM types detected!&#39;, call. = FALSE)
  }
                          }
 )
)

my_new_prot &lt;- protein$new(seq = &quot;ARNDKLLQWYTTARD&quot;, ptm = 
                   list(
                     site = c(3, 5), 
                     type = c(&#39;phosphorylation&#39;, &#39;methylation&#39;)
                    ))
str(my_new_prot)</code></pre>
<p><strong>Congratulations! You are familiar with S3, S4 and R6 object models by now!</strong></p>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
